{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook to visualize data collected from various databases - DASH\n",
    " \n",
    "**Developed by** :Srivalli Kolla\n",
    "\n",
    "**Created on** : 23 October, 2024\n",
    "\n",
    "**Last modified** : 24 October, 2024\n",
    "\n",
    "**Würzburg Institute for Systems Immunology & Julius-Maximilian-Universität Würzburg**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Env : dash"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dash is a framework for building interactive web applications and dashboards. It integrates Plotly for creating dynamic visualizations that can respond to user inputs. The app layout is defined using HTML components, allowing for customizable and user-friendly interfaces. Callbacks enable real-time interactivity, updating visual elements based on user actions. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dash\n",
    "import dash_core_components as dcc\n",
    "import dash_html_components as html\n",
    "import plotly.express as px\n",
    "import pandas as pd\n",
    "import plotly.graph_objects as go\n",
    "import io\n",
    "import zipfile\n",
    "import os\n",
    "import tempfile\n",
    "\n",
    "from dash import html, dcc, Input, Output, State\n",
    "from plotly.io import write_image\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Load file and remove spaces from column names\n",
    "2. Convert all string columns to lowercase\n",
    "3. Preprocess the DataFrame and inititae DASH app"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data-modality</th>\n",
       "      <th>Protocol</th>\n",
       "      <th>Repository</th>\n",
       "      <th>Study-Title</th>\n",
       "      <th>Dataset name</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Tissue-Broad</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Disease</th>\n",
       "      <th>Developmentalstage</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Cell count</th>\n",
       "      <th>Sex(female,male,unknown)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sc,sn</td>\n",
       "      <td>scRNA, snRNA</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Cells of the adult human heart</td>\n",
       "      <td>All — Cells of the adult human heart</td>\n",
       "      <td>Human</td>\n",
       "      <td>Apex,LA,LV,RA,RV,IVseptum</td>\n",
       "      <td>apex of heart,left cardiac atrium,heart left v...</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>European,Asian</td>\n",
       "      <td>486134</td>\n",
       "      <td>219695;266439;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Spatial multi-omic map of human myocardial inf...</td>\n",
       "      <td>All-snRNA-Spatial multi-omic map of human myoc...</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>European</td>\n",
       "      <td>41663</td>\n",
       "      <td>10682;30981;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Spatial multi-omic map of human myocardial inf...</td>\n",
       "      <td>All-snRNA-Spatial multi-omic map of human myoc...</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>Myocardial infarction</td>\n",
       "      <td>Adult</td>\n",
       "      <td>European</td>\n",
       "      <td>150132</td>\n",
       "      <td>45354;104778;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sc</td>\n",
       "      <td>scRNA</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Spatially resolved multiomics of human cardiac...</td>\n",
       "      <td>Combined single cell and single nuclei RNA-Seq...</td>\n",
       "      <td>Human</td>\n",
       "      <td>RA</td>\n",
       "      <td>right cardiac atrium</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>European</td>\n",
       "      <td>1</td>\n",
       "      <td>1;0;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sc</td>\n",
       "      <td>scRNA</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Construction of a human cell landscape at sing...</td>\n",
       "      <td>Construction of a human cell landscape at sing...</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult,Fetal</td>\n",
       "      <td>Han Chinese</td>\n",
       "      <td>10783</td>\n",
       "      <td>9475;1308;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Pathogenic variants damage cell composition an...</td>\n",
       "      <td>DCM/ACM heart cell atlas: All cells</td>\n",
       "      <td>Human</td>\n",
       "      <td>RV,LV,IVseptum</td>\n",
       "      <td>heart right ventricle,heart left ventricle,int...</td>\n",
       "      <td>DCM</td>\n",
       "      <td>Adult,Fetal,Pediatric</td>\n",
       "      <td>European</td>\n",
       "      <td>482581</td>\n",
       "      <td>127956;354625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Pathogenic variants damage cell composition an...</td>\n",
       "      <td>DCM/ACM heart cell atlas: All cells</td>\n",
       "      <td>Human</td>\n",
       "      <td>RV,LV,IVseptum</td>\n",
       "      <td>heart right ventricle,heart left ventricle,int...</td>\n",
       "      <td>Arrhythmogenic right ventricular cardiomyopathy</td>\n",
       "      <td>Adult,Fetal,Pediatric</td>\n",
       "      <td>European</td>\n",
       "      <td>104496</td>\n",
       "      <td>34322;70174;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Pathogenic variants damage cell composition an...</td>\n",
       "      <td>DCM/ACM heart cell atlas: All cells</td>\n",
       "      <td>Human</td>\n",
       "      <td>RV,LV,IVseptum</td>\n",
       "      <td>heart right ventricle,heart left ventricle,int...</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult,Fetal,Pediatric</td>\n",
       "      <td>European</td>\n",
       "      <td>67246</td>\n",
       "      <td>16932;50314;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Pathogenic variants damage cell composition an...</td>\n",
       "      <td>DCM/ACM heart cell atlas: All cells</td>\n",
       "      <td>Human</td>\n",
       "      <td>RV,LV,IVseptum</td>\n",
       "      <td>heart right ventricle,heart left ventricle,int...</td>\n",
       "      <td>Non-compaction cardiomyopathy</td>\n",
       "      <td>Adult,Fetal,Pediatric</td>\n",
       "      <td>European</td>\n",
       "      <td>11632</td>\n",
       "      <td>11632;0;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>sc</td>\n",
       "      <td>scRNA</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Integrated adult and foetal heart single-cell ...</td>\n",
       "      <td>Integrated adult and foetal hearts</td>\n",
       "      <td>Human</td>\n",
       "      <td>Apex,basal_zone</td>\n",
       "      <td>apex of heart, basal zone of heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Fetal</td>\n",
       "      <td>unknown</td>\n",
       "      <td>30889</td>\n",
       "      <td>21700;9189;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Spatial multi-omic map of human myocardial inf...</td>\n",
       "      <td>Ischemia-snRNA-Spatial multi-omic map of human...</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>Myocardial infarction</td>\n",
       "      <td>Adult</td>\n",
       "      <td>European</td>\n",
       "      <td>19722</td>\n",
       "      <td>0;19722;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Single-nucleus cross-tissue molecular referenc...</td>\n",
       "      <td>Single-nucleus cross-tissue molecular referenc...</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>anterior wall of left ventricle</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>unknown</td>\n",
       "      <td>36574</td>\n",
       "      <td>24507;12067;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>sn</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>A human cell atlas of fetal gene expression</td>\n",
       "      <td>Survey of human embryonic development</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Fetal</td>\n",
       "      <td>unknown</td>\n",
       "      <td>101749</td>\n",
       "      <td>55878;45871;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>sc</td>\n",
       "      <td>scRNA</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>Tabula Sapiens - All Cells</td>\n",
       "      <td>Human</td>\n",
       "      <td>Atria,Ventricle,Coronary_artery</td>\n",
       "      <td>cardiac atrium,cardiac ventricle,coronary artery</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Hispanic or Latin American,European</td>\n",
       "      <td>16372</td>\n",
       "      <td>0;16372;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>HBM236.JPVT.769</td>\n",
       "      <td>HBM236.JPVT.769</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>5907</td>\n",
       "      <td>0;5907;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>sc</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>HBM296.KZXD.676</td>\n",
       "      <td>HBM296.KZXD.676</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>6924</td>\n",
       "      <td>6924;0;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>sc</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>HBM342.CMHT.948</td>\n",
       "      <td>HBM342.CMHT.948</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>179</td>\n",
       "      <td>179;0;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>sc</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>HBM364.XLVH.839</td>\n",
       "      <td>HBM364.XLVH.839</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>14262</td>\n",
       "      <td>0;14262;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>sc</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>HBM385.LQVK.975</td>\n",
       "      <td>HBM385.LQVK.975</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>10720</td>\n",
       "      <td>0;10720;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>HBM537.MVDQ.934</td>\n",
       "      <td>HBM537.MVDQ.934</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>2996</td>\n",
       "      <td>0;2996;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>sc</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>HBM563.ZTQV.233</td>\n",
       "      <td>HBM563.ZTQV.233</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>9542</td>\n",
       "      <td>0;9542;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>sc</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>HBM573.MWDW.928</td>\n",
       "      <td>HBM573.MWDW.928</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>9176</td>\n",
       "      <td>0;9176;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>sc</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>HBM592.DXHG.266</td>\n",
       "      <td>HBM592.DXHG.266</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>14553</td>\n",
       "      <td>0;14553;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>sc</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>HBM599.GNJN.777</td>\n",
       "      <td>HBM599.GNJN.777</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>10570</td>\n",
       "      <td>10570;0;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>sc</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>HBM852.VXGN.375</td>\n",
       "      <td>HBM852.VXGN.375</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>10404</td>\n",
       "      <td>0;10404;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>sc</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>HBM927.DZCV.762</td>\n",
       "      <td>HBM927.DZCV.762</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>7233</td>\n",
       "      <td>0;7233;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>sc</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>HBM943.DFKN.947</td>\n",
       "      <td>HBM943.DFKN.947</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>12845</td>\n",
       "      <td>0;12845;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>HBM957.TXXZ.387</td>\n",
       "      <td>HBM957.TXXZ.387</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>668</td>\n",
       "      <td>668;0;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>sc</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>HBM982.THGM.772</td>\n",
       "      <td>HBM982.THGM.772</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>10478</td>\n",
       "      <td>0;10478;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Single nucleus RNA sequencing data from human ...</td>\n",
       "      <td>PRJEB59734</td>\n",
       "      <td>Human</td>\n",
       "      <td>RV</td>\n",
       "      <td>Heart right ventricle</td>\n",
       "      <td>Cardiac transplant</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1635</td>\n",
       "      <td>222;1413;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Single nucleus RNA sequencing data from human ...</td>\n",
       "      <td>PRJEB59734</td>\n",
       "      <td>Human</td>\n",
       "      <td>RV,LV</td>\n",
       "      <td>Heart right ventricle, left ventricle</td>\n",
       "      <td>DCM</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1176</td>\n",
       "      <td>0;1176;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Single nucleus RNA sequencing data from human ...</td>\n",
       "      <td>PRJEB59734</td>\n",
       "      <td>Human</td>\n",
       "      <td>RV,LV</td>\n",
       "      <td>Heart right ventricle, left ventricle</td>\n",
       "      <td>Cardiomyopathy</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>2219</td>\n",
       "      <td>436;1783;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Single nucleus RNA sequencing data from human ...</td>\n",
       "      <td>PRJEB59734</td>\n",
       "      <td>Human</td>\n",
       "      <td>RV</td>\n",
       "      <td>Heart right ventricle</td>\n",
       "      <td>Myocarditis</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1064</td>\n",
       "      <td>519;545;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>single-nucleus RNA-seq of human SGs</td>\n",
       "      <td>PRJNA1007964</td>\n",
       "      <td>Human</td>\n",
       "      <td>Sympathetic ganglion</td>\n",
       "      <td>Sympathetic ganglion</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>524</td>\n",
       "      <td>0;524;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Expression profiling by snRNAseq of left ventr...</td>\n",
       "      <td>PRJNA1127309</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>LV</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0</td>\n",
       "      <td>0;0;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Expression profiling by snRNAseq of left ventr...</td>\n",
       "      <td>PRJNA1127309</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>LV</td>\n",
       "      <td>Aorticstenotic</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>153</td>\n",
       "      <td>0;153;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Cellular Atlas of Human Heart Failure</td>\n",
       "      <td>PRJNA762100</td>\n",
       "      <td>Human</td>\n",
       "      <td>LVapex</td>\n",
       "      <td>LV apex</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>African American,White</td>\n",
       "      <td>13455</td>\n",
       "      <td>5089;8366;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>sc+sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Cellular Atlas of Human Heart Failure</td>\n",
       "      <td>PRJNA762100</td>\n",
       "      <td>Human</td>\n",
       "      <td>LVapex</td>\n",
       "      <td>LV apex</td>\n",
       "      <td>DCM</td>\n",
       "      <td>Adult</td>\n",
       "      <td>African American,White</td>\n",
       "      <td>4136</td>\n",
       "      <td>2342;1794;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Defining the fetal gene program at single cell...</td>\n",
       "      <td>PRJNA767653</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>~100mg of left ventricular Heart muscle</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Pediatric</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1231</td>\n",
       "      <td>0;0;1231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Defining the fetal gene program at single cell...</td>\n",
       "      <td>PRJNA767653</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>~100mg of left ventricular Heart muscle</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>938</td>\n",
       "      <td>551;387;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Defining the fetal gene program at single cell...</td>\n",
       "      <td>PRJNA767653</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>~100mg of left ventricular Heart muscle</td>\n",
       "      <td>DCM</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1792</td>\n",
       "      <td>837;955;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Transcriptional and immune landscape of cardia...</td>\n",
       "      <td>PRJNA847259</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV,Apex</td>\n",
       "      <td>LV, Apex</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>194</td>\n",
       "      <td>0;194;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Single-cell genomics improves the discovery of...</td>\n",
       "      <td>PRJNA934594</td>\n",
       "      <td>Human</td>\n",
       "      <td>RV,LV,IVseptum,Apex</td>\n",
       "      <td>Right ventricle,Interventricular septum,Apex,L...</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>3139</td>\n",
       "      <td>0;3139;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Defining cardiac functional recovery in end-st...</td>\n",
       "      <td>PRJNA939636</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>LV apex</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>22119</td>\n",
       "      <td>4204;21237;2880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>sc</td>\n",
       "      <td>scRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Spatially resolved multiomics of human cardiac...</td>\n",
       "      <td>PRJEB61891</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV,RV</td>\n",
       "      <td>Heart left ventricle, RV</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Caucassian</td>\n",
       "      <td>0</td>\n",
       "      <td>0;0;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Human Obstructive and Non-obstructive Hypertro...</td>\n",
       "      <td>PRJNA753316</td>\n",
       "      <td>Human</td>\n",
       "      <td>Septum,Myectomy</td>\n",
       "      <td>Adult human septal tissue,Adult human myectomy...</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0</td>\n",
       "      <td>0;0;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Human Obstructive and Non-obstructive Hypertro...</td>\n",
       "      <td>PRJNA753316</td>\n",
       "      <td>Human</td>\n",
       "      <td>Septum,Myectomy</td>\n",
       "      <td>Adult human septal tissue,Adult human myectomy...</td>\n",
       "      <td>Obstructive HCM</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0</td>\n",
       "      <td>0;0;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Human Obstructive and Non-obstructive Hypertro...</td>\n",
       "      <td>PRJNA753316</td>\n",
       "      <td>Human</td>\n",
       "      <td>Septum,Myectomy</td>\n",
       "      <td>Adult human septal tissue,Adult human myectomy...</td>\n",
       "      <td>Non-Obstructive HCM</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0</td>\n",
       "      <td>0;0;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Single Nuclei RNA Sequencing Analysis of Cardi...</td>\n",
       "      <td>PRJNA769125</td>\n",
       "      <td>Human</td>\n",
       "      <td>RV</td>\n",
       "      <td>Heart (right ventricle)</td>\n",
       "      <td>Covid-19</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Hispanic</td>\n",
       "      <td>208</td>\n",
       "      <td>65;143;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>sn</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>The Macrophage Landscape Across the Lifespan o...</td>\n",
       "      <td>PRJNA936884</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>Left ventricle</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0</td>\n",
       "      <td>0;0;0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Data-modality      Protocol Repository  \\\n",
       "0          sc,sn  scRNA, snRNA  CellxGene   \n",
       "1             sn         snRNA  CellxGene   \n",
       "2             sn         snRNA  CellxGene   \n",
       "3             sc         scRNA  CellxGene   \n",
       "4             sc         scRNA  CellxGene   \n",
       "5             sn         snRNA  CellxGene   \n",
       "6             sn         snRNA  CellxGene   \n",
       "7             sn         snRNA  CellxGene   \n",
       "8             sn         snRNA  CellxGene   \n",
       "9             sc         scRNA  CellxGene   \n",
       "10            sn         snRNA  CellxGene   \n",
       "11            sn         snRNA  CellxGene   \n",
       "12            sn   sci-RNA-seq  CellxGene   \n",
       "13            sc         scRNA  CellxGene   \n",
       "14            sn         snRNA     Hubmap   \n",
       "15            sc   sci-RNA-seq     Hubmap   \n",
       "16            sc   sci-RNA-seq     Hubmap   \n",
       "17            sc   sci-RNA-seq     Hubmap   \n",
       "18            sc   sci-RNA-seq     Hubmap   \n",
       "19            sn         snRNA     Hubmap   \n",
       "20            sc   sci-RNA-seq     Hubmap   \n",
       "21            sc   sci-RNA-seq     Hubmap   \n",
       "22            sc   sci-RNA-seq     Hubmap   \n",
       "23            sc   sci-RNA-seq     Hubmap   \n",
       "24            sc   sci-RNA-seq     Hubmap   \n",
       "25            sc   sci-RNA-seq     Hubmap   \n",
       "26            sc   sci-RNA-seq     Hubmap   \n",
       "27            sn         snRNA     Hubmap   \n",
       "28            sc   sci-RNA-seq     Hubmap   \n",
       "29            sn         snRNA        SRA   \n",
       "30            sn         snRNA        SRA   \n",
       "31            sn         snRNA        SRA   \n",
       "32            sn         snRNA        SRA   \n",
       "33            sn         snRNA        SRA   \n",
       "34            sn         snRNA        SRA   \n",
       "35            sn         snRNA        SRA   \n",
       "36            sn         snRNA        SRA   \n",
       "37         sc+sn         snRNA        SRA   \n",
       "38            sn         snRNA        SRA   \n",
       "39            sn         snRNA        SRA   \n",
       "40            sn         snRNA        SRA   \n",
       "41            sn         snRNA        SRA   \n",
       "42            sn         snRNA        SRA   \n",
       "43            sn         snRNA        SRA   \n",
       "44            sc         scRNA        SRA   \n",
       "45            sn         snRNA        SRA   \n",
       "46            sn         snRNA        SRA   \n",
       "47            sn         snRNA        SRA   \n",
       "48            sn         snRNA        SRA   \n",
       "49            sn         snRNA        SRA   \n",
       "\n",
       "                                          Study-Title  \\\n",
       "0                      Cells of the adult human heart   \n",
       "1   Spatial multi-omic map of human myocardial inf...   \n",
       "2   Spatial multi-omic map of human myocardial inf...   \n",
       "3   Spatially resolved multiomics of human cardiac...   \n",
       "4   Construction of a human cell landscape at sing...   \n",
       "5   Pathogenic variants damage cell composition an...   \n",
       "6   Pathogenic variants damage cell composition an...   \n",
       "7   Pathogenic variants damage cell composition an...   \n",
       "8   Pathogenic variants damage cell composition an...   \n",
       "9   Integrated adult and foetal heart single-cell ...   \n",
       "10  Spatial multi-omic map of human myocardial inf...   \n",
       "11  Single-nucleus cross-tissue molecular referenc...   \n",
       "12        A human cell atlas of fetal gene expression   \n",
       "13                                     Tabula Sapiens   \n",
       "14                                    HBM236.JPVT.769   \n",
       "15                                    HBM296.KZXD.676   \n",
       "16                                    HBM342.CMHT.948   \n",
       "17                                    HBM364.XLVH.839   \n",
       "18                                    HBM385.LQVK.975   \n",
       "19                                    HBM537.MVDQ.934   \n",
       "20                                    HBM563.ZTQV.233   \n",
       "21                                    HBM573.MWDW.928   \n",
       "22                                    HBM592.DXHG.266   \n",
       "23                                    HBM599.GNJN.777   \n",
       "24                                    HBM852.VXGN.375   \n",
       "25                                    HBM927.DZCV.762   \n",
       "26                                    HBM943.DFKN.947   \n",
       "27                                    HBM957.TXXZ.387   \n",
       "28                                    HBM982.THGM.772   \n",
       "29  Single nucleus RNA sequencing data from human ...   \n",
       "30  Single nucleus RNA sequencing data from human ...   \n",
       "31  Single nucleus RNA sequencing data from human ...   \n",
       "32  Single nucleus RNA sequencing data from human ...   \n",
       "33                single-nucleus RNA-seq of human SGs   \n",
       "34  Expression profiling by snRNAseq of left ventr...   \n",
       "35  Expression profiling by snRNAseq of left ventr...   \n",
       "36              Cellular Atlas of Human Heart Failure   \n",
       "37              Cellular Atlas of Human Heart Failure   \n",
       "38  Defining the fetal gene program at single cell...   \n",
       "39  Defining the fetal gene program at single cell...   \n",
       "40  Defining the fetal gene program at single cell...   \n",
       "41  Transcriptional and immune landscape of cardia...   \n",
       "42  Single-cell genomics improves the discovery of...   \n",
       "43  Defining cardiac functional recovery in end-st...   \n",
       "44  Spatially resolved multiomics of human cardiac...   \n",
       "45  Human Obstructive and Non-obstructive Hypertro...   \n",
       "46  Human Obstructive and Non-obstructive Hypertro...   \n",
       "47  Human Obstructive and Non-obstructive Hypertro...   \n",
       "48  Single Nuclei RNA Sequencing Analysis of Cardi...   \n",
       "49  The Macrophage Landscape Across the Lifespan o...   \n",
       "\n",
       "                                         Dataset name Organism  \\\n",
       "0                All — Cells of the adult human heart    Human   \n",
       "1   All-snRNA-Spatial multi-omic map of human myoc...    Human   \n",
       "2   All-snRNA-Spatial multi-omic map of human myoc...    Human   \n",
       "3   Combined single cell and single nuclei RNA-Seq...    Human   \n",
       "4   Construction of a human cell landscape at sing...    Human   \n",
       "5                 DCM/ACM heart cell atlas: All cells    Human   \n",
       "6                 DCM/ACM heart cell atlas: All cells    Human   \n",
       "7                 DCM/ACM heart cell atlas: All cells    Human   \n",
       "8                 DCM/ACM heart cell atlas: All cells    Human   \n",
       "9                  Integrated adult and foetal hearts    Human   \n",
       "10  Ischemia-snRNA-Spatial multi-omic map of human...    Human   \n",
       "11  Single-nucleus cross-tissue molecular referenc...    Human   \n",
       "12              Survey of human embryonic development    Human   \n",
       "13                        Tabula Sapiens - All Cells     Human   \n",
       "14                                    HBM236.JPVT.769    Human   \n",
       "15                                    HBM296.KZXD.676    Human   \n",
       "16                                    HBM342.CMHT.948    Human   \n",
       "17                                    HBM364.XLVH.839    Human   \n",
       "18                                    HBM385.LQVK.975    Human   \n",
       "19                                    HBM537.MVDQ.934    Human   \n",
       "20                                    HBM563.ZTQV.233    Human   \n",
       "21                                    HBM573.MWDW.928    Human   \n",
       "22                                    HBM592.DXHG.266    Human   \n",
       "23                                    HBM599.GNJN.777    Human   \n",
       "24                                    HBM852.VXGN.375    Human   \n",
       "25                                    HBM927.DZCV.762    Human   \n",
       "26                                    HBM943.DFKN.947    Human   \n",
       "27                                    HBM957.TXXZ.387    Human   \n",
       "28                                    HBM982.THGM.772    Human   \n",
       "29                                         PRJEB59734    Human   \n",
       "30                                         PRJEB59734    Human   \n",
       "31                                         PRJEB59734    Human   \n",
       "32                                         PRJEB59734    Human   \n",
       "33                                       PRJNA1007964    Human   \n",
       "34                                       PRJNA1127309    Human   \n",
       "35                                       PRJNA1127309    Human   \n",
       "36                                        PRJNA762100    Human   \n",
       "37                                        PRJNA762100    Human   \n",
       "38                                        PRJNA767653    Human   \n",
       "39                                        PRJNA767653    Human   \n",
       "40                                        PRJNA767653    Human   \n",
       "41                                        PRJNA847259    Human   \n",
       "42                                        PRJNA934594    Human   \n",
       "43                                        PRJNA939636    Human   \n",
       "44                                         PRJEB61891    Human   \n",
       "45                                        PRJNA753316    Human   \n",
       "46                                        PRJNA753316    Human   \n",
       "47                                        PRJNA753316    Human   \n",
       "48                                        PRJNA769125    Human   \n",
       "49                                        PRJNA936884    Human   \n",
       "\n",
       "                       Tissue-Broad  \\\n",
       "0         Apex,LA,LV,RA,RV,IVseptum   \n",
       "1                                LV   \n",
       "2                                LV   \n",
       "3                                RA   \n",
       "4                             Heart   \n",
       "5                    RV,LV,IVseptum   \n",
       "6                    RV,LV,IVseptum   \n",
       "7                    RV,LV,IVseptum   \n",
       "8                    RV,LV,IVseptum   \n",
       "9                   Apex,basal_zone   \n",
       "10                               LV   \n",
       "11                               LV   \n",
       "12                            Heart   \n",
       "13  Atria,Ventricle,Coronary_artery   \n",
       "14                            Heart   \n",
       "15                            Heart   \n",
       "16                            Heart   \n",
       "17                            Heart   \n",
       "18                            Heart   \n",
       "19                            Heart   \n",
       "20                            Heart   \n",
       "21                            Heart   \n",
       "22                            Heart   \n",
       "23                            Heart   \n",
       "24                            Heart   \n",
       "25                            Heart   \n",
       "26                            Heart   \n",
       "27                            Heart   \n",
       "28                            Heart   \n",
       "29                               RV   \n",
       "30                            RV,LV   \n",
       "31                            RV,LV   \n",
       "32                               RV   \n",
       "33             Sympathetic ganglion   \n",
       "34                               LV   \n",
       "35                               LV   \n",
       "36                           LVapex   \n",
       "37                           LVapex   \n",
       "38                               LV   \n",
       "39                               LV   \n",
       "40                               LV   \n",
       "41                          LV,Apex   \n",
       "42              RV,LV,IVseptum,Apex   \n",
       "43                               LV   \n",
       "44                            LV,RV   \n",
       "45                  Septum,Myectomy   \n",
       "46                  Septum,Myectomy   \n",
       "47                  Septum,Myectomy   \n",
       "48                               RV   \n",
       "49                               LV   \n",
       "\n",
       "                                               Tissue  \\\n",
       "0   apex of heart,left cardiac atrium,heart left v...   \n",
       "1                                heart left ventricle   \n",
       "2                                heart left ventricle   \n",
       "3                                right cardiac atrium   \n",
       "4                                               heart   \n",
       "5   heart right ventricle,heart left ventricle,int...   \n",
       "6   heart right ventricle,heart left ventricle,int...   \n",
       "7   heart right ventricle,heart left ventricle,int...   \n",
       "8   heart right ventricle,heart left ventricle,int...   \n",
       "9                  apex of heart, basal zone of heart   \n",
       "10                               heart left ventricle   \n",
       "11                    anterior wall of left ventricle   \n",
       "12                                              heart   \n",
       "13   cardiac atrium,cardiac ventricle,coronary artery   \n",
       "14                                              Heart   \n",
       "15                                              Heart   \n",
       "16                                              Heart   \n",
       "17                                              Heart   \n",
       "18                                              Heart   \n",
       "19                                              Heart   \n",
       "20                                              Heart   \n",
       "21                                              Heart   \n",
       "22                                              Heart   \n",
       "23                                              Heart   \n",
       "24                                              Heart   \n",
       "25                                              Heart   \n",
       "26                                              Heart   \n",
       "27                                              Heart   \n",
       "28                                              Heart   \n",
       "29                              Heart right ventricle   \n",
       "30              Heart right ventricle, left ventricle   \n",
       "31              Heart right ventricle, left ventricle   \n",
       "32                              Heart right ventricle   \n",
       "33                               Sympathetic ganglion   \n",
       "34                                                 LV   \n",
       "35                                                 LV   \n",
       "36                                            LV apex   \n",
       "37                                            LV apex   \n",
       "38            ~100mg of left ventricular Heart muscle   \n",
       "39            ~100mg of left ventricular Heart muscle   \n",
       "40            ~100mg of left ventricular Heart muscle   \n",
       "41                                           LV, Apex   \n",
       "42  Right ventricle,Interventricular septum,Apex,L...   \n",
       "43                                            LV apex   \n",
       "44                           Heart left ventricle, RV   \n",
       "45  Adult human septal tissue,Adult human myectomy...   \n",
       "46  Adult human septal tissue,Adult human myectomy...   \n",
       "47  Adult human septal tissue,Adult human myectomy...   \n",
       "48                            Heart (right ventricle)   \n",
       "49                                     Left ventricle   \n",
       "\n",
       "                                            Disease     Developmentalstage  \\\n",
       "0                                            Normal                  Adult   \n",
       "1                                            Normal                  Adult   \n",
       "2                             Myocardial infarction                  Adult   \n",
       "3                                            Normal                  Adult   \n",
       "4                                            Normal            Adult,Fetal   \n",
       "5                                               DCM  Adult,Fetal,Pediatric   \n",
       "6   Arrhythmogenic right ventricular cardiomyopathy  Adult,Fetal,Pediatric   \n",
       "7                                            Normal  Adult,Fetal,Pediatric   \n",
       "8                     Non-compaction cardiomyopathy  Adult,Fetal,Pediatric   \n",
       "9                                            Normal                  Fetal   \n",
       "10                            Myocardial infarction                  Adult   \n",
       "11                                           Normal                  Adult   \n",
       "12                                           Normal                  Fetal   \n",
       "13                                           Normal                  Adult   \n",
       "14                                           Normal                  Adult   \n",
       "15                                           Normal                  Adult   \n",
       "16                                           Normal                  Adult   \n",
       "17                                           Normal                  Adult   \n",
       "18                                           Normal                  Adult   \n",
       "19                                           Normal                  Adult   \n",
       "20                                           Normal                  Adult   \n",
       "21                                           Normal                  Adult   \n",
       "22                                           Normal                  Adult   \n",
       "23                                           Normal                  Adult   \n",
       "24                                           Normal                  Adult   \n",
       "25                                           Normal                  Adult   \n",
       "26                                           Normal                  Adult   \n",
       "27                                           Normal                  Adult   \n",
       "28                                           Normal                  Adult   \n",
       "29                               Cardiac transplant                  Adult   \n",
       "30                                              DCM                  Adult   \n",
       "31                                   Cardiomyopathy                  Adult   \n",
       "32                                      Myocarditis                  Adult   \n",
       "33                                           Normal                  Adult   \n",
       "34                                           Normal                  Adult   \n",
       "35                                   Aorticstenotic                  Adult   \n",
       "36                                           Normal                  Adult   \n",
       "37                                              DCM                  Adult   \n",
       "38                                           Normal              Pediatric   \n",
       "39                                           Normal                  Adult   \n",
       "40                                              DCM                  Adult   \n",
       "41                                           Normal                  Adult   \n",
       "42                                           Normal                  Adult   \n",
       "43                                           Normal                  Adult   \n",
       "44                                           Normal                  Adult   \n",
       "45                                           Normal                  Adult   \n",
       "46                                  Obstructive HCM                  Adult   \n",
       "47                              Non-Obstructive HCM                  Adult   \n",
       "48                                         Covid-19                  Adult   \n",
       "49                                           Normal                  Adult   \n",
       "\n",
       "                              Ethnicity  Cell count Sex(female,male,unknown)  \n",
       "0                        European,Asian      486134          219695;266439;0  \n",
       "1                              European       41663            10682;30981;0  \n",
       "2                              European      150132           45354;104778;0  \n",
       "3                              European           1                    1;0;0  \n",
       "4                           Han Chinese       10783              9475;1308;0  \n",
       "5                              European      482581            127956;354625  \n",
       "6                              European      104496            34322;70174;0  \n",
       "7                              European       67246            16932;50314;0  \n",
       "8                              European       11632                11632;0;0  \n",
       "9                               unknown       30889             21700;9189;0  \n",
       "10                             European       19722                0;19722;0  \n",
       "11                              unknown       36574            24507;12067;0  \n",
       "12                              unknown      101749            55878;45871;0  \n",
       "13  Hispanic or Latin American,European       16372                0;16372;0  \n",
       "14                                White        5907                 0;5907;0  \n",
       "15                                White        6924                 6924;0;0  \n",
       "16                                White         179                  179;0;0  \n",
       "17                                White       14262                0;14262;0  \n",
       "18                                White       10720                0;10720;0  \n",
       "19                                White        2996                 0;2996;0  \n",
       "20                                White        9542                 0;9542;0  \n",
       "21                                White        9176                 0;9176;0  \n",
       "22                                White       14553                0;14553;0  \n",
       "23                                White       10570                10570;0;0  \n",
       "24                                White       10404                0;10404;0  \n",
       "25                                White        7233                 0;7233;0  \n",
       "26                                White       12845                0;12845;0  \n",
       "27            Black or African American         668                  668;0;0  \n",
       "28            Black or African American       10478                0;10478;0  \n",
       "29                              Unknown        1635               222;1413;0  \n",
       "30                              Unknown        1176                 0;1176;0  \n",
       "31                              Unknown        2219               436;1783;0  \n",
       "32                              Unknown        1064                519;545;0  \n",
       "33                              Unknown         524                  0;524;0  \n",
       "34                              Unknown           0                    0;0;0  \n",
       "35                              Unknown         153                  0;153;0  \n",
       "36               African American,White       13455              5089;8366;0  \n",
       "37               African American,White        4136              2342;1794;0  \n",
       "38                              Unknown        1231                 0;0;1231  \n",
       "39                              Unknown         938                551;387;0  \n",
       "40                              Unknown        1792                837;955;0  \n",
       "41                                White         194                  0;194;0  \n",
       "42                                White        3139                 0;3139;0  \n",
       "43                              Unknown       22119          4204;21237;2880  \n",
       "44                           Caucassian           0                    0;0;0  \n",
       "45                              Unknown           0                    0;0;0  \n",
       "46                              Unknown           0                    0;0;0  \n",
       "47                              Unknown           0                    0;0;0  \n",
       "48                             Hispanic         208                 65;143;0  \n",
       "49                              Unknown           0                    0;0;0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('../data_visualization/heart_cellular_collectives.csv')\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pre Processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Create a copy of the DataFrame to avoid modifying the original\n",
    "2. Function to parse sex values as seperate to add during merging and apply on the column \n",
    "3. Function to safely convert to numeric during merging and apply on numerical columns in our Dataset name\n",
    "4. Cartedully define the numeric and string columns and apply defined functions\n",
    "5. Aggregate(Merge) Numerical columns\n",
    "6. Define a function to caredully hanndle multiple data during merging by converting them into a comma sperated list without duplicates\n",
    "7. Now Aggregate categorical columns and define to use function from 6 to handle properly\n",
    "8. Now combine numerical and categorical aggregated data into one Dataset name\n",
    "9. Ensure that Sex column is properly demonstrated\n",
    "10. Add desired order for columns for output aand also appennd any additional columns which are missing and sort according to order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess_and_aggregate(df):\n",
    "    df_copy = df.copy()\n",
    "\n",
    "    # Function to parse the 'Sex(female,male,unknown)' column into female, male, and unknown counts\n",
    "    def parse_sex(x):\n",
    "        if pd.isna(x) or x == '':\n",
    "            return 0, 0, 0\n",
    "        parts = str(x).split(';')\n",
    "        if len(parts) != 3:\n",
    "            return 0, 0, 0\n",
    "        return (\n",
    "            int(parts[0]) if parts[0].isdigit() else 0,\n",
    "            int(parts[1]) if parts[1].isdigit() else 0,\n",
    "            int(parts[2]) if parts[2].isdigit() else 0\n",
    "        )\n",
    "\n",
    "    # Apply parsing function and create 'female', 'male', and 'unknown' columns\n",
    "    df_copy[['female', 'male', 'unknown']] = df_copy['Sex(female,male,unknown)'].apply(lambda x: pd.Series(parse_sex(x)))\n",
    "\n",
    "    # Function to safely convert to numeric values\n",
    "    def safe_numeric(x):\n",
    "        return pd.to_numeric(x, errors='coerce').fillna(0)\n",
    "\n",
    "    # Ensure 'Cell count' is numeric\n",
    "    df_copy['Cell count'] = safe_numeric(df_copy['Cell count'])\n",
    "\n",
    "    # Columns to aggregate\n",
    "    numerical_cols = ['Cell count', 'female', 'male', 'unknown']\n",
    "    string_cols = [col for col in df_copy.columns if col not in numerical_cols + ['Study-Title', 'Dataset name']]\n",
    "\n",
    "    # Aggregate numerical columns by sum\n",
    "    num_agg = df_copy.groupby(['Study-Title', 'Dataset name'])[numerical_cols].sum().reset_index()\n",
    "\n",
    "    # Function to join unique, comma-separated values for string columns\n",
    "    def join_unique_comma_separated(series):\n",
    "        combined_list = []\n",
    "        for value in series.dropna().astype(str):\n",
    "            combined_list.extend(value.split(','))\n",
    "        return ','.join(sorted(set(filter(None, combined_list))))\n",
    "\n",
    "    # Aggregate string columns by joining unique values\n",
    "    str_agg = df_copy.groupby(['Study-Title', 'Dataset name'])[string_cols].agg(join_unique_comma_separated).reset_index()\n",
    "\n",
    "    # Merge numerical and string aggregations\n",
    "    grouped_df = pd.merge(num_agg, str_agg, on=['Study-Title', 'Dataset name'])\n",
    "\n",
    "    # Reconstruct the 'Sex(female,male,unknown)' column\n",
    "    grouped_df['Sex(female,male,unknown)'] = (\n",
    "        grouped_df['female'].astype(int).astype(str) + ';' +\n",
    "        grouped_df['male'].astype(int).astype(str) + ';' +\n",
    "        grouped_df['unknown'].astype(int).astype(str)\n",
    "    )\n",
    "    \n",
    "    # Drop the temporary 'female', 'male', and 'unknown' columns\n",
    "    grouped_df = grouped_df.drop(columns=['female', 'male', 'unknown'])\n",
    "\n",
    "    # Define the desired columns and ensure there are no duplicates\n",
    "    desired_columns = [\n",
    "        'Study-Title', 'Dataset name', 'Data-modality', 'Repository',\n",
    "        'Organism', 'Tissue-Broad', 'Tissue', 'Disease',\n",
    "        'Developmentalstage', 'Ethnicity', 'Sex(female,male,unknown)', 'Cell count'\n",
    "    ]\n",
    "\n",
    "    # Add any additional columns that are not in the desired list\n",
    "    additional_columns = [col for col in grouped_df.columns if col not in desired_columns]\n",
    "    column_order = desired_columns + additional_columns\n",
    "\n",
    "    # Reorder the DataFrame columns\n",
    "    grouped_df = grouped_df[column_order]\n",
    "\n",
    "    return grouped_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study-Title</th>\n",
       "      <th>Dataset name</th>\n",
       "      <th>Data-modality</th>\n",
       "      <th>Repository</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Tissue-Broad</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Disease</th>\n",
       "      <th>Developmentalstage</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Sex(female,male,unknown)</th>\n",
       "      <th>Cell count</th>\n",
       "      <th>Protocol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A human cell atlas of fetal gene expression</td>\n",
       "      <td>Survey of human embryonic development</td>\n",
       "      <td>sn</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Fetal</td>\n",
       "      <td>unknown</td>\n",
       "      <td>55878;45871;0</td>\n",
       "      <td>101749</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cells of the adult human heart</td>\n",
       "      <td>All — Cells of the adult human heart</td>\n",
       "      <td>sc,sn</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Human</td>\n",
       "      <td>Apex,IVseptum,LA,LV,RA,RV</td>\n",
       "      <td>apex of heart,heart left ventricle,heart right...</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Asian,European</td>\n",
       "      <td>219695;266439;0</td>\n",
       "      <td>486134</td>\n",
       "      <td>snRNA,scRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cellular Atlas of Human Heart Failure</td>\n",
       "      <td>PRJNA762100</td>\n",
       "      <td>sc+sn,sn</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Human</td>\n",
       "      <td>LVapex</td>\n",
       "      <td>LV apex</td>\n",
       "      <td>DCM,Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>African American,White</td>\n",
       "      <td>7431;10160;0</td>\n",
       "      <td>17591</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Construction of a human cell landscape at sing...</td>\n",
       "      <td>Construction of a human cell landscape at sing...</td>\n",
       "      <td>sc</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult,Fetal</td>\n",
       "      <td>Han Chinese</td>\n",
       "      <td>9475;1308;0</td>\n",
       "      <td>10783</td>\n",
       "      <td>scRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Defining cardiac functional recovery in end-st...</td>\n",
       "      <td>PRJNA939636</td>\n",
       "      <td>sn</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>LV apex</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>4204;21237;2880</td>\n",
       "      <td>22119</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Defining the fetal gene program at single cell...</td>\n",
       "      <td>PRJNA767653</td>\n",
       "      <td>sn</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>~100mg of left ventricular Heart muscle</td>\n",
       "      <td>DCM,Normal</td>\n",
       "      <td>Adult,Pediatric</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1388;1342;1231</td>\n",
       "      <td>3961</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Expression profiling by snRNAseq of left ventr...</td>\n",
       "      <td>PRJNA1127309</td>\n",
       "      <td>sn</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>LV</td>\n",
       "      <td>Aorticstenotic,Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0;153;0</td>\n",
       "      <td>153</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HBM236.JPVT.769</td>\n",
       "      <td>HBM236.JPVT.769</td>\n",
       "      <td>sn</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>0;5907;0</td>\n",
       "      <td>5907</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HBM296.KZXD.676</td>\n",
       "      <td>HBM296.KZXD.676</td>\n",
       "      <td>sc</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>6924;0;0</td>\n",
       "      <td>6924</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HBM342.CMHT.948</td>\n",
       "      <td>HBM342.CMHT.948</td>\n",
       "      <td>sc</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>179;0;0</td>\n",
       "      <td>179</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HBM364.XLVH.839</td>\n",
       "      <td>HBM364.XLVH.839</td>\n",
       "      <td>sc</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>0;14262;0</td>\n",
       "      <td>14262</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HBM385.LQVK.975</td>\n",
       "      <td>HBM385.LQVK.975</td>\n",
       "      <td>sc</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>0;10720;0</td>\n",
       "      <td>10720</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HBM537.MVDQ.934</td>\n",
       "      <td>HBM537.MVDQ.934</td>\n",
       "      <td>sn</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>0;2996;0</td>\n",
       "      <td>2996</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HBM563.ZTQV.233</td>\n",
       "      <td>HBM563.ZTQV.233</td>\n",
       "      <td>sc</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>0;9542;0</td>\n",
       "      <td>9542</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>HBM573.MWDW.928</td>\n",
       "      <td>HBM573.MWDW.928</td>\n",
       "      <td>sc</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>0;9176;0</td>\n",
       "      <td>9176</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>HBM592.DXHG.266</td>\n",
       "      <td>HBM592.DXHG.266</td>\n",
       "      <td>sc</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>0;14553;0</td>\n",
       "      <td>14553</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HBM599.GNJN.777</td>\n",
       "      <td>HBM599.GNJN.777</td>\n",
       "      <td>sc</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>10570;0;0</td>\n",
       "      <td>10570</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HBM852.VXGN.375</td>\n",
       "      <td>HBM852.VXGN.375</td>\n",
       "      <td>sc</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>0;10404;0</td>\n",
       "      <td>10404</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>HBM927.DZCV.762</td>\n",
       "      <td>HBM927.DZCV.762</td>\n",
       "      <td>sc</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>0;7233;0</td>\n",
       "      <td>7233</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>HBM943.DFKN.947</td>\n",
       "      <td>HBM943.DFKN.947</td>\n",
       "      <td>sc</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>0;12845;0</td>\n",
       "      <td>12845</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>HBM957.TXXZ.387</td>\n",
       "      <td>HBM957.TXXZ.387</td>\n",
       "      <td>sn</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>668;0;0</td>\n",
       "      <td>668</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>HBM982.THGM.772</td>\n",
       "      <td>HBM982.THGM.772</td>\n",
       "      <td>sc</td>\n",
       "      <td>Hubmap</td>\n",
       "      <td>Human</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>0;10478;0</td>\n",
       "      <td>10478</td>\n",
       "      <td>sci-RNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Human Obstructive and Non-obstructive Hypertro...</td>\n",
       "      <td>PRJNA753316</td>\n",
       "      <td>sn</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Human</td>\n",
       "      <td>Myectomy,Septum</td>\n",
       "      <td>Adult human myectomy tissue,Adult human septal...</td>\n",
       "      <td>Non-Obstructive HCM,Normal,Obstructive HCM</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>0</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Integrated adult and foetal heart single-cell ...</td>\n",
       "      <td>Integrated adult and foetal hearts</td>\n",
       "      <td>sc</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Human</td>\n",
       "      <td>Apex,basal_zone</td>\n",
       "      <td>basal zone of heart,apex of heart</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Fetal</td>\n",
       "      <td>unknown</td>\n",
       "      <td>21700;9189;0</td>\n",
       "      <td>30889</td>\n",
       "      <td>scRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Pathogenic variants damage cell composition an...</td>\n",
       "      <td>DCM/ACM heart cell atlas: All cells</td>\n",
       "      <td>sn</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Human</td>\n",
       "      <td>IVseptum,LV,RV</td>\n",
       "      <td>heart left ventricle,heart right ventricle,int...</td>\n",
       "      <td>Arrhythmogenic right ventricular cardiomyopath...</td>\n",
       "      <td>Adult,Fetal,Pediatric</td>\n",
       "      <td>European</td>\n",
       "      <td>62886;120488;0</td>\n",
       "      <td>665955</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Single Nuclei RNA Sequencing Analysis of Cardi...</td>\n",
       "      <td>PRJNA769125</td>\n",
       "      <td>sn</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Human</td>\n",
       "      <td>RV</td>\n",
       "      <td>Heart (right ventricle)</td>\n",
       "      <td>Covid-19</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Hispanic</td>\n",
       "      <td>65;143;0</td>\n",
       "      <td>208</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Single nucleus RNA sequencing data from human ...</td>\n",
       "      <td>PRJEB59734</td>\n",
       "      <td>sn</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV,RV</td>\n",
       "      <td>left ventricle,Heart right ventricle</td>\n",
       "      <td>Cardiac transplant,Cardiomyopathy,DCM,Myocarditis</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1177;4917;0</td>\n",
       "      <td>6094</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Single-cell genomics improves the discovery of...</td>\n",
       "      <td>PRJNA934594</td>\n",
       "      <td>sn</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Human</td>\n",
       "      <td>Apex,IVseptum,LV,RV</td>\n",
       "      <td>Apex,Interventricular septum,Left ventricle,Ri...</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>0;3139;0</td>\n",
       "      <td>3139</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Single-nucleus cross-tissue molecular referenc...</td>\n",
       "      <td>Single-nucleus cross-tissue molecular referenc...</td>\n",
       "      <td>sn</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>anterior wall of left ventricle</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>unknown</td>\n",
       "      <td>24507;12067;0</td>\n",
       "      <td>36574</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Spatial multi-omic map of human myocardial inf...</td>\n",
       "      <td>All-snRNA-Spatial multi-omic map of human myoc...</td>\n",
       "      <td>sn</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>Myocardial infarction,Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>European</td>\n",
       "      <td>56036;135759;0</td>\n",
       "      <td>191795</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Spatial multi-omic map of human myocardial inf...</td>\n",
       "      <td>Ischemia-snRNA-Spatial multi-omic map of human...</td>\n",
       "      <td>sn</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>Myocardial infarction</td>\n",
       "      <td>Adult</td>\n",
       "      <td>European</td>\n",
       "      <td>0;19722;0</td>\n",
       "      <td>19722</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Spatially resolved multiomics of human cardiac...</td>\n",
       "      <td>Combined single cell and single nuclei RNA-Seq...</td>\n",
       "      <td>sc</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Human</td>\n",
       "      <td>RA</td>\n",
       "      <td>right cardiac atrium</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>European</td>\n",
       "      <td>1;0;0</td>\n",
       "      <td>1</td>\n",
       "      <td>scRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Spatially resolved multiomics of human cardiac...</td>\n",
       "      <td>PRJEB61891</td>\n",
       "      <td>sc</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV,RV</td>\n",
       "      <td>RV,Heart left ventricle</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Caucassian</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>0</td>\n",
       "      <td>scRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>Tabula Sapiens - All Cells</td>\n",
       "      <td>sc</td>\n",
       "      <td>CellxGene</td>\n",
       "      <td>Human</td>\n",
       "      <td>Atria,Coronary_artery,Ventricle</td>\n",
       "      <td>cardiac atrium,cardiac ventricle,coronary artery</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>European,Hispanic or Latin American</td>\n",
       "      <td>0;16372;0</td>\n",
       "      <td>16372</td>\n",
       "      <td>scRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>The Macrophage Landscape Across the Lifespan o...</td>\n",
       "      <td>PRJNA936884</td>\n",
       "      <td>sn</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Human</td>\n",
       "      <td>LV</td>\n",
       "      <td>Left ventricle</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>0</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Transcriptional and immune landscape of cardia...</td>\n",
       "      <td>PRJNA847259</td>\n",
       "      <td>sn</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Human</td>\n",
       "      <td>Apex,LV</td>\n",
       "      <td>Apex,LV</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>White</td>\n",
       "      <td>0;194;0</td>\n",
       "      <td>194</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>single-nucleus RNA-seq of human SGs</td>\n",
       "      <td>PRJNA1007964</td>\n",
       "      <td>sn</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Human</td>\n",
       "      <td>Sympathetic ganglion</td>\n",
       "      <td>Sympathetic ganglion</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0;524;0</td>\n",
       "      <td>524</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Study-Title  \\\n",
       "0         A human cell atlas of fetal gene expression   \n",
       "1                      Cells of the adult human heart   \n",
       "2               Cellular Atlas of Human Heart Failure   \n",
       "3   Construction of a human cell landscape at sing...   \n",
       "4   Defining cardiac functional recovery in end-st...   \n",
       "5   Defining the fetal gene program at single cell...   \n",
       "6   Expression profiling by snRNAseq of left ventr...   \n",
       "7                                     HBM236.JPVT.769   \n",
       "8                                     HBM296.KZXD.676   \n",
       "9                                     HBM342.CMHT.948   \n",
       "10                                    HBM364.XLVH.839   \n",
       "11                                    HBM385.LQVK.975   \n",
       "12                                    HBM537.MVDQ.934   \n",
       "13                                    HBM563.ZTQV.233   \n",
       "14                                    HBM573.MWDW.928   \n",
       "15                                    HBM592.DXHG.266   \n",
       "16                                    HBM599.GNJN.777   \n",
       "17                                    HBM852.VXGN.375   \n",
       "18                                    HBM927.DZCV.762   \n",
       "19                                    HBM943.DFKN.947   \n",
       "20                                    HBM957.TXXZ.387   \n",
       "21                                    HBM982.THGM.772   \n",
       "22  Human Obstructive and Non-obstructive Hypertro...   \n",
       "23  Integrated adult and foetal heart single-cell ...   \n",
       "24  Pathogenic variants damage cell composition an...   \n",
       "25  Single Nuclei RNA Sequencing Analysis of Cardi...   \n",
       "26  Single nucleus RNA sequencing data from human ...   \n",
       "27  Single-cell genomics improves the discovery of...   \n",
       "28  Single-nucleus cross-tissue molecular referenc...   \n",
       "29  Spatial multi-omic map of human myocardial inf...   \n",
       "30  Spatial multi-omic map of human myocardial inf...   \n",
       "31  Spatially resolved multiomics of human cardiac...   \n",
       "32  Spatially resolved multiomics of human cardiac...   \n",
       "33                                     Tabula Sapiens   \n",
       "34  The Macrophage Landscape Across the Lifespan o...   \n",
       "35  Transcriptional and immune landscape of cardia...   \n",
       "36                single-nucleus RNA-seq of human SGs   \n",
       "\n",
       "                                         Dataset name Data-modality  \\\n",
       "0               Survey of human embryonic development            sn   \n",
       "1                All — Cells of the adult human heart         sc,sn   \n",
       "2                                         PRJNA762100      sc+sn,sn   \n",
       "3   Construction of a human cell landscape at sing...            sc   \n",
       "4                                         PRJNA939636            sn   \n",
       "5                                         PRJNA767653            sn   \n",
       "6                                        PRJNA1127309            sn   \n",
       "7                                     HBM236.JPVT.769            sn   \n",
       "8                                     HBM296.KZXD.676            sc   \n",
       "9                                     HBM342.CMHT.948            sc   \n",
       "10                                    HBM364.XLVH.839            sc   \n",
       "11                                    HBM385.LQVK.975            sc   \n",
       "12                                    HBM537.MVDQ.934            sn   \n",
       "13                                    HBM563.ZTQV.233            sc   \n",
       "14                                    HBM573.MWDW.928            sc   \n",
       "15                                    HBM592.DXHG.266            sc   \n",
       "16                                    HBM599.GNJN.777            sc   \n",
       "17                                    HBM852.VXGN.375            sc   \n",
       "18                                    HBM927.DZCV.762            sc   \n",
       "19                                    HBM943.DFKN.947            sc   \n",
       "20                                    HBM957.TXXZ.387            sn   \n",
       "21                                    HBM982.THGM.772            sc   \n",
       "22                                        PRJNA753316            sn   \n",
       "23                 Integrated adult and foetal hearts            sc   \n",
       "24                DCM/ACM heart cell atlas: All cells            sn   \n",
       "25                                        PRJNA769125            sn   \n",
       "26                                         PRJEB59734            sn   \n",
       "27                                        PRJNA934594            sn   \n",
       "28  Single-nucleus cross-tissue molecular referenc...            sn   \n",
       "29  All-snRNA-Spatial multi-omic map of human myoc...            sn   \n",
       "30  Ischemia-snRNA-Spatial multi-omic map of human...            sn   \n",
       "31  Combined single cell and single nuclei RNA-Seq...            sc   \n",
       "32                                         PRJEB61891            sc   \n",
       "33                        Tabula Sapiens - All Cells             sc   \n",
       "34                                        PRJNA936884            sn   \n",
       "35                                        PRJNA847259            sn   \n",
       "36                                       PRJNA1007964            sn   \n",
       "\n",
       "   Repository Organism                     Tissue-Broad  \\\n",
       "0   CellxGene    Human                            Heart   \n",
       "1   CellxGene    Human        Apex,IVseptum,LA,LV,RA,RV   \n",
       "2         SRA    Human                           LVapex   \n",
       "3   CellxGene    Human                            Heart   \n",
       "4         SRA    Human                               LV   \n",
       "5         SRA    Human                               LV   \n",
       "6         SRA    Human                               LV   \n",
       "7      Hubmap    Human                            Heart   \n",
       "8      Hubmap    Human                            Heart   \n",
       "9      Hubmap    Human                            Heart   \n",
       "10     Hubmap    Human                            Heart   \n",
       "11     Hubmap    Human                            Heart   \n",
       "12     Hubmap    Human                            Heart   \n",
       "13     Hubmap    Human                            Heart   \n",
       "14     Hubmap    Human                            Heart   \n",
       "15     Hubmap    Human                            Heart   \n",
       "16     Hubmap    Human                            Heart   \n",
       "17     Hubmap    Human                            Heart   \n",
       "18     Hubmap    Human                            Heart   \n",
       "19     Hubmap    Human                            Heart   \n",
       "20     Hubmap    Human                            Heart   \n",
       "21     Hubmap    Human                            Heart   \n",
       "22        SRA    Human                  Myectomy,Septum   \n",
       "23  CellxGene    Human                  Apex,basal_zone   \n",
       "24  CellxGene    Human                   IVseptum,LV,RV   \n",
       "25        SRA    Human                               RV   \n",
       "26        SRA    Human                            LV,RV   \n",
       "27        SRA    Human              Apex,IVseptum,LV,RV   \n",
       "28  CellxGene    Human                               LV   \n",
       "29  CellxGene    Human                               LV   \n",
       "30  CellxGene    Human                               LV   \n",
       "31  CellxGene    Human                               RA   \n",
       "32        SRA    Human                            LV,RV   \n",
       "33  CellxGene    Human  Atria,Coronary_artery,Ventricle   \n",
       "34        SRA    Human                               LV   \n",
       "35        SRA    Human                          Apex,LV   \n",
       "36        SRA    Human             Sympathetic ganglion   \n",
       "\n",
       "                                               Tissue  \\\n",
       "0                                               heart   \n",
       "1   apex of heart,heart left ventricle,heart right...   \n",
       "2                                             LV apex   \n",
       "3                                               heart   \n",
       "4                                             LV apex   \n",
       "5             ~100mg of left ventricular Heart muscle   \n",
       "6                                                  LV   \n",
       "7                                               Heart   \n",
       "8                                               Heart   \n",
       "9                                               Heart   \n",
       "10                                              Heart   \n",
       "11                                              Heart   \n",
       "12                                              Heart   \n",
       "13                                              Heart   \n",
       "14                                              Heart   \n",
       "15                                              Heart   \n",
       "16                                              Heart   \n",
       "17                                              Heart   \n",
       "18                                              Heart   \n",
       "19                                              Heart   \n",
       "20                                              Heart   \n",
       "21                                              Heart   \n",
       "22  Adult human myectomy tissue,Adult human septal...   \n",
       "23                  basal zone of heart,apex of heart   \n",
       "24  heart left ventricle,heart right ventricle,int...   \n",
       "25                            Heart (right ventricle)   \n",
       "26               left ventricle,Heart right ventricle   \n",
       "27  Apex,Interventricular septum,Left ventricle,Ri...   \n",
       "28                    anterior wall of left ventricle   \n",
       "29                               heart left ventricle   \n",
       "30                               heart left ventricle   \n",
       "31                               right cardiac atrium   \n",
       "32                            RV,Heart left ventricle   \n",
       "33   cardiac atrium,cardiac ventricle,coronary artery   \n",
       "34                                     Left ventricle   \n",
       "35                                            Apex,LV   \n",
       "36                               Sympathetic ganglion   \n",
       "\n",
       "                                              Disease     Developmentalstage  \\\n",
       "0                                              Normal                  Fetal   \n",
       "1                                              Normal                  Adult   \n",
       "2                                          DCM,Normal                  Adult   \n",
       "3                                              Normal            Adult,Fetal   \n",
       "4                                              Normal                  Adult   \n",
       "5                                          DCM,Normal        Adult,Pediatric   \n",
       "6                               Aorticstenotic,Normal                  Adult   \n",
       "7                                              Normal                  Adult   \n",
       "8                                              Normal                  Adult   \n",
       "9                                              Normal                  Adult   \n",
       "10                                             Normal                  Adult   \n",
       "11                                             Normal                  Adult   \n",
       "12                                             Normal                  Adult   \n",
       "13                                             Normal                  Adult   \n",
       "14                                             Normal                  Adult   \n",
       "15                                             Normal                  Adult   \n",
       "16                                             Normal                  Adult   \n",
       "17                                             Normal                  Adult   \n",
       "18                                             Normal                  Adult   \n",
       "19                                             Normal                  Adult   \n",
       "20                                             Normal                  Adult   \n",
       "21                                             Normal                  Adult   \n",
       "22         Non-Obstructive HCM,Normal,Obstructive HCM                  Adult   \n",
       "23                                             Normal                  Fetal   \n",
       "24  Arrhythmogenic right ventricular cardiomyopath...  Adult,Fetal,Pediatric   \n",
       "25                                           Covid-19                  Adult   \n",
       "26  Cardiac transplant,Cardiomyopathy,DCM,Myocarditis                  Adult   \n",
       "27                                             Normal                  Adult   \n",
       "28                                             Normal                  Adult   \n",
       "29                       Myocardial infarction,Normal                  Adult   \n",
       "30                              Myocardial infarction                  Adult   \n",
       "31                                             Normal                  Adult   \n",
       "32                                             Normal                  Adult   \n",
       "33                                             Normal                  Adult   \n",
       "34                                             Normal                  Adult   \n",
       "35                                             Normal                  Adult   \n",
       "36                                             Normal                  Adult   \n",
       "\n",
       "                              Ethnicity Sex(female,male,unknown)  Cell count  \\\n",
       "0                               unknown            55878;45871;0      101749   \n",
       "1                        Asian,European          219695;266439;0      486134   \n",
       "2                African American,White             7431;10160;0       17591   \n",
       "3                           Han Chinese              9475;1308;0       10783   \n",
       "4                               Unknown          4204;21237;2880       22119   \n",
       "5                               Unknown           1388;1342;1231        3961   \n",
       "6                               Unknown                  0;153;0         153   \n",
       "7                                 White                 0;5907;0        5907   \n",
       "8                                 White                 6924;0;0        6924   \n",
       "9                                 White                  179;0;0         179   \n",
       "10                                White                0;14262;0       14262   \n",
       "11                                White                0;10720;0       10720   \n",
       "12                                White                 0;2996;0        2996   \n",
       "13                                White                 0;9542;0        9542   \n",
       "14                                White                 0;9176;0        9176   \n",
       "15                                White                0;14553;0       14553   \n",
       "16                                White                10570;0;0       10570   \n",
       "17                                White                0;10404;0       10404   \n",
       "18                                White                 0;7233;0        7233   \n",
       "19                                White                0;12845;0       12845   \n",
       "20            Black or African American                  668;0;0         668   \n",
       "21            Black or African American                0;10478;0       10478   \n",
       "22                              Unknown                    0;0;0           0   \n",
       "23                              unknown             21700;9189;0       30889   \n",
       "24                             European           62886;120488;0      665955   \n",
       "25                             Hispanic                 65;143;0         208   \n",
       "26                              Unknown              1177;4917;0        6094   \n",
       "27                                White                 0;3139;0        3139   \n",
       "28                              unknown            24507;12067;0       36574   \n",
       "29                             European           56036;135759;0      191795   \n",
       "30                             European                0;19722;0       19722   \n",
       "31                             European                    1;0;0           1   \n",
       "32                           Caucassian                    0;0;0           0   \n",
       "33  European,Hispanic or Latin American                0;16372;0       16372   \n",
       "34                              Unknown                    0;0;0           0   \n",
       "35                                White                  0;194;0         194   \n",
       "36                              Unknown                  0;524;0         524   \n",
       "\n",
       "        Protocol  \n",
       "0    sci-RNA-seq  \n",
       "1    snRNA,scRNA  \n",
       "2          snRNA  \n",
       "3          scRNA  \n",
       "4          snRNA  \n",
       "5          snRNA  \n",
       "6          snRNA  \n",
       "7          snRNA  \n",
       "8    sci-RNA-seq  \n",
       "9    sci-RNA-seq  \n",
       "10   sci-RNA-seq  \n",
       "11   sci-RNA-seq  \n",
       "12         snRNA  \n",
       "13   sci-RNA-seq  \n",
       "14   sci-RNA-seq  \n",
       "15   sci-RNA-seq  \n",
       "16   sci-RNA-seq  \n",
       "17   sci-RNA-seq  \n",
       "18   sci-RNA-seq  \n",
       "19   sci-RNA-seq  \n",
       "20         snRNA  \n",
       "21   sci-RNA-seq  \n",
       "22         snRNA  \n",
       "23         scRNA  \n",
       "24         snRNA  \n",
       "25         snRNA  \n",
       "26         snRNA  \n",
       "27         snRNA  \n",
       "28         snRNA  \n",
       "29         snRNA  \n",
       "30         snRNA  \n",
       "31         scRNA  \n",
       "32         scRNA  \n",
       "33         scRNA  \n",
       "34         snRNA  \n",
       "35         snRNA  \n",
       "36         snRNA  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped_df = preprocess_and_aggregate(df)\n",
    "grouped_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Function to get unique values from comma-separated columns\n",
    "2. Function to select the data from comma seperated column if it matches with any of the value from selection\n",
    "3. Function to split comma-separated values and explode into multiple rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique_comma_separated_values(df, column):\n",
    "    unique_values = set()\n",
    "    for values in df[column].dropna():\n",
    "        for value in values.split(','):\n",
    "            unique_values.add(value.strip())\n",
    "    return sorted(unique_values)\n",
    "\n",
    "def any_in_string(selected_values, string_value):\n",
    "    if not string_value or not selected_values:\n",
    "        return False\n",
    "    string_list = [s.strip() for s in string_value.split(',')]\n",
    "\n",
    "    return any(value in string_list for value in selected_values)\n",
    "\n",
    "def split_and_explode(df, col):\n",
    "    return df.assign(**{col: df[col].str.split(',')}).explode(col)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_diseases = get_unique_comma_separated_values(grouped_df, 'Disease')\n",
    "unique_tissue_broad = get_unique_comma_separated_values(grouped_df, 'Tissue-Broad')\n",
    "unique_modalities = get_unique_comma_separated_values(grouped_df, 'Data-modality')\n",
    "unique_repositories = grouped_df['Repository'].unique().tolist() "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DASH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "app = dash.Dash(__name__)\n",
    "\n",
    "app.layout = html.Div([\n",
    "    html.H1(\"Heart Data Consortium Visualizations\", style={'textAlign': 'center'}),\n",
    "    \n",
    "    # Filters\n",
    "    html.Div([\n",
    "        html.H2(\"Disease Filters\"),\n",
    "        dcc.Dropdown(\n",
    "            id='disease-dropdown',\n",
    "            options=[{'label': disease, 'value': disease} for disease in unique_diseases],\n",
    "            value=unique_diseases,\n",
    "            multi=True,\n",
    "            placeholder=\"Select Disease\"\n",
    "        ),\n",
    "    ], style={'margin-bottom': '20px'}),\n",
    "\n",
    "    html.Div([\n",
    "        html.H2(\"Tissue-Broad Filters\"),\n",
    "        dcc.Dropdown(\n",
    "            id='Tissue-Broad-dropdown',\n",
    "            options=[{'label': tissue, 'value': tissue} for tissue in unique_tissue_broad],\n",
    "            value=unique_tissue_broad,\n",
    "            multi=True,\n",
    "            placeholder=\"Select Tissue-Broad\"\n",
    "        ),\n",
    "    ], style={'margin-bottom': '20px'}),\n",
    "\n",
    "    html.Div([\n",
    "        html.H2(\"Data Modality Filters\"),\n",
    "        dcc.Dropdown(\n",
    "            id='modality-dropdown',\n",
    "            options=[{'label': modality, 'value': modality} for modality in unique_modalities],\n",
    "            value=unique_modalities,\n",
    "            multi=True,\n",
    "            placeholder=\"Select Data Modality\"\n",
    "        ),\n",
    "    ], style={'margin-bottom': '20px'}),\n",
    "\n",
    "    html.Div([\n",
    "        html.H2(\"Repository Filters\"),\n",
    "        dcc.Dropdown(\n",
    "            id='repository-dropdown',\n",
    "            options=[{'label': repo, 'value': repo} for repo in unique_repositories],\n",
    "            value=unique_repositories,\n",
    "            multi=True,\n",
    "            placeholder=\"Select Repository\"\n",
    "        ),\n",
    "    ], style={'margin-bottom': '20px'}),\n",
    "\n",
    "    # Graphs\n",
    "    dcc.Graph(id='repository-piechart'),\n",
    "    dcc.Graph(id='repository-barplot'),\n",
    "    dcc.Graph(id='disease-barplot'),\n",
    "    dcc.Graph(id='disease-pie-chart'),\n",
    "    dcc.Graph(id='cell-count-barplot'),\n",
    "    dcc.Graph(id='tissue-broad-barchart'),\n",
    "    dcc.Graph(id='tissue-broad-piechart'),\n",
    "    dcc.Graph(id='modality-piechart'),\n",
    "    dcc.Graph(id='developmental-stage-barplot'),\n",
    "    dcc.Graph(id='developmental-stage-pie-chart'),\n",
    "    dcc.Graph(id='sex-distribution-barplot-counts'),\n",
    "    dcc.Graph(id='sex-distribution-barplot-percentage'),\n",
    "    dcc.Graph(id='sex-distribution-pie-chart'),\n",
    "    dcc.Graph(id='ethnicity-barplot-counts'),  \n",
    "    dcc.Graph(id='ethnicity-barplot-percentage'),\n",
    "    dcc.Graph(id='ethnicity-pie-chart'),\n",
    "    dcc.Graph(id='treemap-ethnicity'),\n",
    "    dcc.Graph(id ='treemap-total'),\n",
    "    \n",
    "    # Save Button\n",
    "    html.Div([\n",
    "        html.H3(\"Specify Save Path\"),\n",
    "        dcc.Input(id='save-path-input', type='text', value='./Plots/', style={'width': '100%', 'margin-bottom': '20px'}),\n",
    "        html.Button(\"Save All Figures\", id=\"save-button\", n_clicks=0),\n",
    "        html.Div(id=\"save-status\", style={'margin-top': '20px', 'font-weight': 'bold'})\n",
    "    ]),\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Callback to save all graphs as PNG images to local path\n",
    "@app.callback(\n",
    "    Output('save-status', 'children'),\n",
    "    [Input('save-button', 'n_clicks')],\n",
    "    [State('save-path-input', 'value'),\n",
    "     State('repository-piechart', 'figure'),\n",
    "     State('repository-barplot', 'figure'),\n",
    "     State('disease-barplot', 'figure'),\n",
    "     State('disease-pie-chart', 'figure'),\n",
    "     State('cell-count-barplot', 'figure'),\n",
    "     State('modality-piechart', 'figure'),\n",
    "     State('developmental-stage-barplot', 'figure'),\n",
    "     State('developmental-stage-pie-chart', 'figure'),\n",
    "     State('sex-distribution-barplot-counts', 'figure'),\n",
    "     State('sex-distribution-barplot-percentage', 'figure'),\n",
    "     State('sex-distribution-pie-chart', 'figure'),\n",
    "     State('ethnicity-barplot-counts', 'figure'),\n",
    "     State('ethnicity-barplot-percentage', 'figure'),\n",
    "     State('ethnicity-pie-chart', 'figure'),\n",
    "     State('treemap-ethnicity', 'figure'),\n",
    "     State('treemap-total','figure')]\n",
    "     \n",
    ")\n",
    "def save_all_figures(n_clicks, save_path, *figures):\n",
    "    if n_clicks == 0:\n",
    "        return dash.no_update  # Do not save until the button is clicked\n",
    "\n",
    "    if not save_path.endswith('/'):\n",
    "        save_path += '/'\n",
    "    \n",
    "    # Create the directory if it does not exist\n",
    "    os.makedirs(save_path, exist_ok=True)\n",
    "\n",
    "    # Save each figure as a PNG file\n",
    "    for i, fig_dict in enumerate(figures):\n",
    "        if fig_dict is not None:\n",
    "            fig = go.Figure(fig_dict)  # Convert dict to Figure object\n",
    "            fig_name = f\"figure_{i}.png\"\n",
    "            fig_path = os.path.join(save_path, fig_name)\n",
    "            try:\n",
    "                # Update layout to include legend outside the plot, vertically aligned (Top-right corner)\n",
    "                fig.update_layout(\n",
    "                    autosize=False,  # Disable autosizing to manually control layout\n",
    "                    width=1500,      # Increase width to give space for legend\n",
    "                    height=1000,     # Increase height for more room\n",
    "                    margin=dict(l=50, r=200, t=50, b=50),  # Add larger right margin for the legend\n",
    "                    legend=dict(\n",
    "                        orientation='v',    # Vertical legend\n",
    "                        yanchor='top',      # Anchor the legend to the top\n",
    "                        y=1,                # Position the legend at the top\n",
    "                        xanchor='right',    # Anchor the legend to the right\n",
    "                        x=1.2               # Position the legend outside the plot (right of the graph)\n",
    "                    )\n",
    "                )\n",
    "                # Save the figure as a PNG with 300 DPI\n",
    "                write_image(fig, file=fig_path, format='png', scale=3)\n",
    "            except Exception as e:\n",
    "                print(f\"Error saving figure {i}: {e}\")\n",
    "    \n",
    "    return \"All figures saved successfully!\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Filter counts based on selection\n",
    "2. Convert 'Cell count' to numeric and handle any errors\n",
    "3. Aggregate the total cell counts by repository\n",
    "4. Plotting pie chart and bar graph\n",
    "5. Adding text in graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Callback for Repository bar plot and pie chart\n",
    "@app.callback(\n",
    "    [Output('repository-piechart', 'figure'),\n",
    "     Output('repository-barplot', 'figure')],\n",
    "    [Input('disease-dropdown', 'value'),\n",
    "     Input('Tissue-Broad-dropdown', 'value'),\n",
    "     Input('modality-dropdown', 'value'),\n",
    "     Input('repository-dropdown', 'value')]\n",
    ")\n",
    "def update_repository_charts(selected_diseases, selected_tissue, selected_modalities, selected_repositories):\n",
    "    \n",
    "    # Fill NaN values with an empty string to avoid issues\n",
    "    df_copy = df.copy()\n",
    "    df_copy['Disease'] = df_copy['Disease'].fillna('')\n",
    "    df_copy['Tissue-Broad'] = df_copy['Tissue-Broad'].fillna('')\n",
    "    df_copy['Data-modality'] = df_copy['Data-modality'].fillna('')\n",
    "    df_copy['Repository'] = df_copy['Repository'].fillna('')\n",
    "\n",
    "    # Apply filtering only if selections are made; otherwise, skip that filter\n",
    "    filtered_df = df_copy[\n",
    "        (df_copy['Disease'].apply(lambda x: any_in_string(selected_diseases, x)) if selected_diseases else True) &\n",
    "        (df_copy['Tissue-Broad'].apply(lambda x: any_in_string(selected_tissue, x)) if selected_tissue else True) &\n",
    "        (df_copy['Data-modality'].apply(lambda x: any_in_string(selected_modalities, x)) if selected_modalities else True) &\n",
    "        (df_copy['Repository'].apply(lambda x: any_in_string(selected_repositories, x)) if selected_repositories else True)\n",
    "    ]\n",
    "\n",
    "    # Check if the filtered DataFrame is empty\n",
    "    if filtered_df.empty:\n",
    "        empty_fig = px.pie(title=\"No data available for selected filters.\")\n",
    "        return empty_fig, empty_fig\n",
    "\n",
    "    # Aggregate cell counts by repository\n",
    "    repo_agg_df = filtered_df.groupby('Repository')['Cell count'].sum().reset_index()\n",
    "    total_cell_count = repo_agg_df['Cell count'].sum()\n",
    "\n",
    "    # Pie chart for repository distribution\n",
    "    pie_fig = px.pie(repo_agg_df, names='Repository', values='Cell count',\n",
    "                     labels={'Repository': 'Repository', 'Cell count': 'Number of Cells'},\n",
    "                     title=f\"Distribution of Cells by Repository (Total: {total_cell_count:,})\",\n",
    "                     hole=0.3)\n",
    "\n",
    "    # Bar plot for cell counts by repository\n",
    "    bar_fig = px.bar(repo_agg_df, x='Repository', y='Cell count',\n",
    "                     labels={'Repository': 'Repository', 'Cell count': 'Number of Cells'},\n",
    "                     title=f\"Number of Cells by Repository (Total: {total_cell_count:,})\",\n",
    "                     text='Cell count')\n",
    "\n",
    "    # Update traces for better visualization\n",
    "    pie_fig.update_traces(textposition='inside', textinfo='percent')\n",
    "    bar_fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')\n",
    "\n",
    "    return pie_fig, bar_fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Callback for disease Bar plot and Pie chart\n",
    "@app.callback(\n",
    "    [Output('disease-pie-chart', 'figure'),\n",
    "     Output('disease-barplot', 'figure')],\n",
    "    [Input('disease-dropdown', 'value'),\n",
    "     Input('Tissue-Broad-dropdown', 'value'),\n",
    "     Input('modality-dropdown', 'value'),\n",
    "     Input('repository-dropdown', 'value')]\n",
    ")\n",
    "def update_disease_charts(selected_diseases, selected_tissue, selected_modalities, selected_repositories):\n",
    "\n",
    "    # Fill NaN values with an empty string for filtering\n",
    "    df_copy = df.copy()\n",
    "    df_copy['Disease'] = df_copy['Disease'].fillna('')\n",
    "    df_copy['Tissue-Broad'] = df_copy['Tissue-Broad'].fillna('')\n",
    "    df_copy['Data-modality'] = df_copy['Data-modality'].fillna('')\n",
    "    df_copy['Repository'] = df_copy['Repository'].fillna('')\n",
    "\n",
    "    # Apply conditional filtering based on the selected dropdown values\n",
    "    filtered_df = df_copy[\n",
    "        (df_copy['Disease'].apply(lambda x: any_in_string(selected_diseases, x)) if selected_diseases else True) &\n",
    "        (df_copy['Tissue-Broad'].apply(lambda x: any_in_string(selected_tissue, x)) if selected_tissue else True) &\n",
    "        (df_copy['Data-modality'].apply(lambda x: any_in_string(selected_modalities, x)) if selected_modalities else True) &\n",
    "        (df_copy['Repository'].apply(lambda x: any_in_string(selected_repositories, x)) if selected_repositories else True)\n",
    "    ]\n",
    "\n",
    "    # If the filtered DataFrame is empty, return empty figures with a message\n",
    "    if filtered_df.empty:\n",
    "        empty_pie = px.pie(title=\"No data available for selected filters.\")\n",
    "        empty_bar = px.bar(title=\"No data available for selected filters.\")\n",
    "        return empty_pie, empty_bar\n",
    "\n",
    "    # Group the filtered DataFrame by 'Disease' and aggregate 'Cell count'\n",
    "    disease_counts = filtered_df.groupby('Disease')['Cell count'].sum().reset_index()\n",
    "\n",
    "    # Create pie chart for Disease distribution\n",
    "    pie_fig = px.pie(disease_counts, names='Disease', values='Cell count',\n",
    "                     title=\"Distribution of Cells by Disease\")\n",
    "\n",
    "    # Create bar chart for Disease counts\n",
    "    bar_fig = px.bar(disease_counts, x='Disease', y='Cell count',\n",
    "                     labels={'Disease': 'Disease', 'Cell count': 'Number of Cells'},\n",
    "                     title=\"Cell Counts by Disease\",\n",
    "                     text='Cell count')\n",
    "\n",
    "    # Update the trace formatting\n",
    "    pie_fig.update_traces(textposition='inside', textinfo='percent')\n",
    "    bar_fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')\n",
    "\n",
    "    return pie_fig, bar_fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Callback for cell-count histogram\n",
    "@app.callback(\n",
    "    Output('cell-count-barplot', 'figure'),\n",
    "    [Input('disease-dropdown', 'value'),\n",
    "     Input('Tissue-Broad-dropdown', 'value'),\n",
    "     Input('modality-dropdown', 'value'),\n",
    "     Input('repository-dropdown', 'value')]\n",
    ")\n",
    "def update_cell_count_histogram(selected_diseases, selected_tissue, selected_modalities, selected_repositories):\n",
    "    # Ensure inputs are lists for consistent filtering\n",
    "    selected_diseases = selected_diseases if isinstance(selected_diseases, list) else [selected_diseases]\n",
    "    selected_tissue = selected_tissue if isinstance(selected_tissue, list) else [selected_tissue]\n",
    "    selected_modalities = selected_modalities if isinstance(selected_modalities, list) else [selected_modalities]\n",
    "    selected_repositories = selected_repositories if isinstance(selected_repositories, list) else [selected_repositories]\n",
    "\n",
    "    # Copy the DataFrame and fill NaN values for filtering\n",
    "    df_copy = df.copy()\n",
    "    df_copy['Disease'] = df_copy['Disease'].fillna('')\n",
    "    df_copy['Tissue-Broad'] = df_copy['Tissue-Broad'].fillna('')\n",
    "    df_copy['Data-modality'] = df_copy['Data-modality'].fillna('')\n",
    "    df_copy['Repository'] = df_copy['Repository'].fillna('')\n",
    "\n",
    "    # Apply conditional filtering based on selected dropdown values\n",
    "    filtered_df = df_copy[\n",
    "        (df_copy['Disease'].apply(lambda x: any_in_string(selected_diseases, x)) if selected_diseases else True) &\n",
    "        (df_copy['Tissue-Broad'].apply(lambda x: any_in_string(selected_tissue, x)) if selected_tissue else True) &\n",
    "        (df_copy['Data-modality'].apply(lambda x: any_in_string(selected_modalities, x)) if selected_modalities else True) &\n",
    "        (df_copy['Repository'].apply(lambda x: any_in_string(selected_repositories, x)) if selected_repositories else True)\n",
    "    ]\n",
    "\n",
    "    # If the filtered DataFrame is empty, return an empty bar plot with a message\n",
    "    if filtered_df.empty:\n",
    "        return px.bar(title=\"No data available for selected filters.\")\n",
    "\n",
    "    # Define a function to split and sum cell counts\n",
    "    def split_and_sum(cell_counts):\n",
    "        if pd.isna(cell_counts):\n",
    "            return 0\n",
    "        if isinstance(cell_counts, str):\n",
    "            return sum(pd.to_numeric(cell_counts.split(';'), errors='coerce').fillna(0))\n",
    "        elif isinstance(cell_counts, (int, float)):\n",
    "            return cell_counts\n",
    "        return 0\n",
    "\n",
    "    # Calculate the cell counts for each row\n",
    "    filtered_df['Cell count'] = filtered_df['Cell count'].apply(split_and_sum)\n",
    "\n",
    "    # Aggregate cell counts by study title\n",
    "    study_cell_counts = filtered_df.groupby('Study-Title')['Cell count'].sum().reset_index()\n",
    "\n",
    "    # Create a bar plot for the aggregated cell counts\n",
    "    fig = px.bar(study_cell_counts, x='Study-Title', y='Cell count',\n",
    "                 labels={'Study-Title': 'Study Title', 'Cell count': 'Number of Cells'},\n",
    "                 title=\"Histogram of Cell Counts per Study\", text='Cell count')\n",
    "    \n",
    "    fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')\n",
    "    \n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Callback for Modality Pie Chart\n",
    "@app.callback(\n",
    "    Output('modality-piechart', 'figure'),\n",
    "    [Input('disease-dropdown', 'value'),\n",
    "     Input('Tissue-Broad-dropdown', 'value'),\n",
    "     Input('modality-dropdown', 'value'),\n",
    "     Input('repository-dropdown', 'value')]\n",
    ")\n",
    "def update_modality_piechart(selected_diseases, selected_tissue, selected_modalities, selected_repositories):\n",
    "    \n",
    "    # Split and explode relevant columns\n",
    "    exploded_df = df.copy()\n",
    "    for col in ['Disease', 'Tissue-Broad', 'Data-modality', 'Repository']:\n",
    "        exploded_df = split_and_explode(exploded_df, col)\n",
    "    \n",
    "    # Filter the exploded DataFrame based on selected filters\n",
    "    filtered_df = exploded_df[\n",
    "        exploded_df['Disease'].str.strip().isin(selected_diseases) &\n",
    "        exploded_df['Tissue-Broad'].str.strip().isin(selected_tissue) &\n",
    "        exploded_df['Data-modality'].str.strip().isin(selected_modalities) &\n",
    "        exploded_df['Repository'].str.strip().isin(selected_repositories)\n",
    "    ]\n",
    "\n",
    "    # Return a message if no data is available after filtering\n",
    "    if filtered_df.empty:\n",
    "        return px.pie(title=\"No data available for the selected filters.\")\n",
    "\n",
    "    # Aggregate the filtered DataFrame\n",
    "    modality_counts = filtered_df.groupby('Data-modality')['Cell count'].sum().reset_index()\n",
    "\n",
    "    # Create a pie chart figure\n",
    "    fig = px.pie(\n",
    "        modality_counts,\n",
    "        names='Data-modality',\n",
    "        values='Cell count',\n",
    "        title=\"Distribution of Data Modalities\",\n",
    "        hole=0.3,\n",
    "        labels={'Data-modality': 'Data Modality', 'Cell count': 'Number of Cells'}\n",
    "    )\n",
    "\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Callback for Developmentalstage plots\n",
    "@app.callback(\n",
    "    [Output('developmental-stage-barplot', 'figure'),\n",
    "     Output('developmental-stage-pie-chart', 'figure')],\n",
    "    [Input('disease-dropdown', 'value'),\n",
    "     Input('Tissue-Broad-dropdown', 'value'),\n",
    "     Input('modality-dropdown', 'value'),\n",
    "     Input('repository-dropdown', 'value')]\n",
    ")\n",
    "def update_developmental_stage_charts(selected_diseases, selected_tissue, selected_modalities, selected_repositories):\n",
    "    exploded_df = split_and_explode(grouped_df, 'Disease')\n",
    "    exploded_df = split_and_explode(exploded_df, 'Tissue-Broad')\n",
    "    exploded_df = split_and_explode(exploded_df, 'Data-modality')\n",
    "    exploded_df = split_and_explode(exploded_df, 'Repository')\n",
    "\n",
    "    # Handle NaN values by filling with an empty string before applying the filters\n",
    "    filtered_df = exploded_df[\n",
    "        exploded_df['Disease'].fillna('').apply(lambda x: x.strip() in selected_diseases) &\n",
    "        exploded_df['Tissue-Broad'].fillna('').apply(lambda x: x.strip() in selected_tissue) &\n",
    "        exploded_df['Data-modality'].fillna('').apply(lambda x: x.strip() in selected_modalities) &\n",
    "        exploded_df['Repository'].fillna('').apply(lambda x: x.strip() in selected_repositories)\n",
    "    ]\n",
    "\n",
    "    if filtered_df.empty:\n",
    "        return (\n",
    "            px.bar(title=\"No data available for selected filters.\"),\n",
    "            px.pie(title=\"No data available for selected filters.\")\n",
    "        )\n",
    "\n",
    "    expanded_stages_df = split_and_explode(filtered_df, 'Developmentalstage')\n",
    "    expanded_stages_df['Cell count'] = pd.to_numeric(expanded_stages_df['Cell count'], errors='coerce')\n",
    "\n",
    "    stage_counts = expanded_stages_df.groupby(['Study-Title', 'Developmentalstage'])['Cell count'].sum().reset_index()\n",
    "\n",
    "    total_counts = stage_counts.groupby('Study-Title')['Cell count'].sum().reset_index()\n",
    "    total_counts.rename(columns={'Cell count': 'Total Count'}, inplace=True)\n",
    "    stage_counts = stage_counts.merge(total_counts, on='Study-Title')\n",
    "    stage_counts['Percentage'] = (stage_counts['Cell count'] / stage_counts['Total Count']) * 100\n",
    "\n",
    "    bar_fig = px.bar(stage_counts, x='Study-Title', y='Cell count', color='Developmentalstage',\n",
    "                     labels={'Study-Title': 'Study Title', 'Cell count': 'Number of Cells'},\n",
    "                     title=\"Cell Counts by Developmentalstage\",\n",
    "                     barmode='stack', text='Cell count')\n",
    "\n",
    "    percentage_bar_fig = px.bar(stage_counts, x='Study-Title', y='Percentage', color='Developmentalstage',\n",
    "                               labels={'Percentage': 'Percentage (%)', 'Developmentalstage': 'Stage'},\n",
    "                               title=\"Percentage of Cell Counts by Developmentalstage\",\n",
    "                               barmode='stack', text='Percentage')\n",
    "\n",
    "\n",
    "    pie_fig = px.pie(stage_counts, values='Cell count', names='Developmentalstage', \n",
    "                     title=\"Percentage of Cell Counts by Developmentalstage\",\n",
    "                     labels={'Cell count': 'Number of Cells', 'Developmentalstage': 'Stage'},\n",
    "                     hole=0.3)\n",
    "    \n",
    "    bar_fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')\n",
    "    percentage_bar_fig.update_traces(texttemplate='%{text:.1f}%', textposition='outside')\n",
    "    pie_fig.update_traces(textinfo='percent')\n",
    "\n",
    "    return percentage_bar_fig, pie_fig\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Callback for Tissue-Broad Bar Plot and Pie Chart\n",
    "@app.callback(\n",
    "    [Output('tissue-broad-barchart', 'figure'),\n",
    "     Output('tissue-broad-piechart', 'figure')],\n",
    "    [Input('disease-dropdown', 'value'),\n",
    "     Input('Tissue-Broad-dropdown', 'value'),\n",
    "     Input('modality-dropdown', 'value'),\n",
    "     Input('repository-dropdown', 'value')]\n",
    ")\n",
    "def update_tissue_broad_charts(selected_diseases, selected_tissue, selected_modalities, selected_repositories):\n",
    "    \n",
    "    # Split and explode relevant columns\n",
    "    exploded_df = df.copy()\n",
    "    for col in ['Disease', 'Tissue-Broad', 'Data-modality', 'Repository']:\n",
    "        exploded_df = split_and_explode(exploded_df, col)\n",
    "    \n",
    "    # Filter the exploded DataFrame based on selected filters\n",
    "    filtered_df = exploded_df[\n",
    "        exploded_df['Disease'].str.strip().isin(selected_diseases) &\n",
    "        exploded_df['Tissue-Broad'].str.strip().isin(selected_tissue) &\n",
    "        exploded_df['Data-modality'].str.strip().isin(selected_modalities) &\n",
    "        exploded_df['Repository'].str.strip().isin(selected_repositories)\n",
    "    ]\n",
    "\n",
    "    # Check if the filtered DataFrame is empty\n",
    "    if filtered_df.empty:\n",
    "        empty_fig = px.bar(title=\"No data available for selected filters.\")\n",
    "        return empty_fig, empty_fig\n",
    "    \n",
    "    # Group the filtered data by Tissue-Broad and sum the Cell counts\n",
    "    tissue_broad_counts = filtered_df.groupby('Tissue-Broad')['Cell count'].sum().reset_index()\n",
    "\n",
    "    # Create the bar plot for Tissue-Broad\n",
    "    bar_fig = px.bar(\n",
    "        tissue_broad_counts,\n",
    "        x='Tissue-Broad',\n",
    "        y='Cell count',\n",
    "        title=\"Tissue-Broad Distribution (Bar Plot)\",\n",
    "        labels={'Cell count': 'Number of Cells'},\n",
    "        text='Cell count'\n",
    "    )\n",
    "    bar_fig.update_layout(xaxis_title='Tissue-Broad', yaxis_title='Cell Count', title_x=0.5)\n",
    "\n",
    "    # Create the pie chart for Tissue-Broad\n",
    "    pie_fig = px.pie(\n",
    "        tissue_broad_counts,\n",
    "        names='Tissue-Broad',\n",
    "        values='Cell count',\n",
    "        title=\"Tissue-Broad Distribution (Pie Chart)\",\n",
    "        labels={'Cell count': 'Number of Cells'},\n",
    "        hole=0.3\n",
    "    )\n",
    "    pie_fig.update_traces(textinfo='percent+label')\n",
    "\n",
    "    return bar_fig, pie_fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Callback for Sex distribution plots\n",
    "@app.callback(\n",
    "    [Output('sex-distribution-barplot-counts', 'figure'),\n",
    "     Output('sex-distribution-barplot-percentage', 'figure'),\n",
    "     Output('sex-distribution-pie-chart', 'figure')],\n",
    "    [Input('disease-dropdown', 'value'),\n",
    "     Input('Tissue-Broad-dropdown', 'value'),\n",
    "     Input('modality-dropdown', 'value'),\n",
    "     Input('repository-dropdown', 'value')]\n",
    ")\n",
    "def update_sex_distribution_charts(selected_diseases, selected_tissue, selected_modalities, selected_repositories):\n",
    "    # Ensure inputs are lists for consistent filtering\n",
    "    selected_diseases = selected_diseases if isinstance(selected_diseases, list) else [selected_diseases]\n",
    "    selected_tissue = selected_tissue if isinstance(selected_tissue, list) else [selected_tissue]\n",
    "    selected_modalities = selected_modalities if isinstance(selected_modalities, list) else [selected_modalities]\n",
    "    selected_repositories = selected_repositories if isinstance(selected_repositories, list) else [selected_repositories]\n",
    "\n",
    "    # Copy the DataFrame and fill NaN values for filtering\n",
    "    df_copy = df.copy()\n",
    "    df_copy['Disease'] = df_copy['Disease'].fillna('')\n",
    "    df_copy['Tissue-Broad'] = df_copy['Tissue-Broad'].fillna('')\n",
    "    df_copy['Data-modality'] = df_copy['Data-modality'].fillna('')\n",
    "    df_copy['Repository'] = df_copy['Repository'].fillna('')\n",
    "    df_copy['Sex(female,male,unknown)'] = df_copy['Sex(female,male,unknown)'].fillna('0;0;0')\n",
    "\n",
    "    # Apply conditional filtering based on selected dropdown values\n",
    "    filtered_df = df_copy[\n",
    "        (df_copy['Disease'].apply(lambda x: any_in_string(selected_diseases, x)) if selected_diseases else True) &\n",
    "        (df_copy['Tissue-Broad'].apply(lambda x: any_in_string(selected_tissue, x)) if selected_tissue else True) &\n",
    "        (df_copy['Data-modality'].apply(lambda x: any_in_string(selected_modalities, x)) if selected_modalities else True) &\n",
    "        (df_copy['Repository'].apply(lambda x: any_in_string(selected_repositories, x)) if selected_repositories else True)\n",
    "    ]\n",
    "\n",
    "    if filtered_df.empty:\n",
    "        # Return empty plots with a message when no data is available\n",
    "        no_data_fig = px.bar(title=\"No data available for selected filters.\")\n",
    "        return no_data_fig, no_data_fig, px.pie(title=\"No data available for selected filters.\")\n",
    "\n",
    "    # Split the 'Sex(female,male,unknown)' column\n",
    "    sex_split = filtered_df['Sex(female,male,unknown)'].str.split(';', expand=True)\n",
    "    filtered_df['Female Count'] = pd.to_numeric(sex_split[0], errors='coerce').fillna(0)\n",
    "    filtered_df['Male Count'] = pd.to_numeric(sex_split[1], errors='coerce').fillna(0)\n",
    "    filtered_df['Unknown Count'] = pd.to_numeric(sex_split[2], errors='coerce').fillna(0)\n",
    "\n",
    "    # Aggregate the data by 'Study-Title'\n",
    "    sex_summary_df = filtered_df.groupby('Study-Title').agg({\n",
    "        'Female Count': 'sum',\n",
    "        'Male Count': 'sum',\n",
    "        'Unknown Count': 'sum'\n",
    "    }).reset_index()\n",
    "\n",
    "    # Creating a stacked bar plot for the counts\n",
    "    fig_counts = px.bar(\n",
    "        sex_summary_df, \n",
    "        x='Study-Title', \n",
    "        y=['Female Count', 'Male Count', 'Unknown Count'],\n",
    "        labels={'value': 'Number of Cells', 'variable': 'Sex'},\n",
    "        title=\"Cell Counts by Sex\", \n",
    "        barmode='stack'\n",
    "    )\n",
    "    fig_counts.update_traces(texttemplate='%{y:.0f}', textposition='outside')\n",
    "    fig_counts.update_layout(\n",
    "        yaxis_title='Cell Count',\n",
    "        title_x=0.5\n",
    "    )\n",
    "\n",
    "    # Calculate percentages for the bar plot\n",
    "    sex_summary_df['Total'] = sex_summary_df[['Female Count', 'Male Count', 'Unknown Count']].sum(axis=1)\n",
    "    sex_summary_df['Female Percentage'] = (sex_summary_df['Female Count'] / sex_summary_df['Total']) * 100\n",
    "    sex_summary_df['Male Percentage'] = (sex_summary_df['Male Count'] / sex_summary_df['Total']) * 100\n",
    "    sex_summary_df['Unknown Percentage'] = (sex_summary_df['Unknown Count'] / sex_summary_df['Total']) * 100\n",
    "\n",
    "    # Creating a stacked bar plot for percentages\n",
    "    fig_percentage = px.bar(\n",
    "        sex_summary_df, \n",
    "        x='Study-Title', \n",
    "        y=['Female Percentage', 'Male Percentage', 'Unknown Percentage'],\n",
    "        labels={'value': 'Percentage (%)', 'variable': 'Sex'},\n",
    "        title=\"Cell Counts Percentage by Sex\", \n",
    "        barmode='stack'\n",
    "    )\n",
    "    fig_percentage.update_traces(texttemplate='%{y:.2f}%', textposition='outside')\n",
    "    fig_percentage.update_layout(\n",
    "        yaxis_title='Percentage (%)',\n",
    "        title_x=0.5\n",
    "    )\n",
    "\n",
    "    # Summing the counts across all studies for the pie chart\n",
    "    total_sex_counts = sex_summary_df[['Female Count', 'Male Count', 'Unknown Count']].sum().reset_index()\n",
    "    total_sex_counts.columns = ['Sex', 'Cell Count']\n",
    "    total_sex_counts['Sex'] = total_sex_counts['Sex'].replace({\n",
    "        'Female Count': 'Female', \n",
    "        'Male Count': 'Male', \n",
    "        'Unknown Count': 'Unknown'\n",
    "    })\n",
    "\n",
    "    # Creating a pie chart for the overall distribution\n",
    "    pie_fig = px.pie(\n",
    "        total_sex_counts, \n",
    "        names='Sex', \n",
    "        values='Cell Count',\n",
    "        title=\"Overall Distribution of Cells by Sex\",\n",
    "        hole=0.3\n",
    "    )\n",
    "    pie_fig.update_traces(textinfo='percent+label')\n",
    "    pie_fig.update_layout(title_x=0.5)\n",
    "\n",
    "    return fig_counts, fig_percentage, pie_fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Callback for Ethnicity Barplot\n",
    "@app.callback(\n",
    "    [Output('ethnicity-barplot-counts', 'figure'),\n",
    "     Output('ethnicity-barplot-percentage', 'figure'),\n",
    "     Output('ethnicity-pie-chart', 'figure')],\n",
    "    [Input('disease-dropdown', 'value'),\n",
    "     Input('Tissue-Broad-dropdown', 'value'),\n",
    "     Input('modality-dropdown', 'value'),\n",
    "     Input('repository-dropdown', 'value')]\n",
    ")\n",
    "def update_ethnicity_charts(selected_diseases, selected_tissue, selected_modalities, selected_repositories):\n",
    "    # Exploding necessary columns to handle multiple entries\n",
    "    exploded_df = split_and_explode(grouped_df, 'Disease')\n",
    "    exploded_df = split_and_explode(exploded_df, 'Tissue-Broad')\n",
    "    exploded_df = split_and_explode(exploded_df, 'Data-modality')\n",
    "    exploded_df = split_and_explode(exploded_df, 'Repository')\n",
    "    exploded_df = split_and_explode(exploded_df, 'Ethnicity')\n",
    "\n",
    "    # Filtering the DataFrame based on the selected filters\n",
    "    filtered_df = exploded_df[\n",
    "        exploded_df['Disease'].apply(lambda x: x.strip() in selected_diseases) &\n",
    "        exploded_df['Tissue-Broad'].apply(lambda x: x.strip() in selected_tissue) &\n",
    "        exploded_df['Data-modality'].apply(lambda x: x.strip() in selected_modalities) &\n",
    "        exploded_df['Repository'].apply(lambda x: x.strip() in selected_repositories)\n",
    "    ]\n",
    "\n",
    "    if filtered_df.empty:\n",
    "        # Return empty plots if no data is available\n",
    "        no_data_fig = px.bar(title=\"No data available for selected filters.\")\n",
    "        return no_data_fig, no_data_fig, px.pie(title=\"No data available for selected filters.\")\n",
    "\n",
    "    # Aggregating the cell count by 'Study-Title' and 'Ethnicity'\n",
    "    ethnicity_counts = filtered_df.groupby(['Study-Title', 'Ethnicity'])['Cell count'].sum().reset_index()\n",
    "\n",
    "    # Creating a bar plot for cell counts by ethnicity\n",
    "    fig_counts = px.bar(\n",
    "        ethnicity_counts, \n",
    "        x='Study-Title', \n",
    "        y='Cell count', \n",
    "        color='Ethnicity',\n",
    "        labels={'Study-Title': 'Study', 'Cell count': 'Number of Cells'},\n",
    "        title=\"Cell Counts by Ethnicity\", \n",
    "        text='Cell count'\n",
    "    )\n",
    "    fig_counts.update_traces(texttemplate='%{text:.0f}', textposition='outside')\n",
    "    fig_counts.update_layout(\n",
    "        yaxis_title='Number of Cells',\n",
    "        title_x=0.5\n",
    "    )\n",
    "\n",
    "    # Calculating the total count for each study to compute percentages\n",
    "    ethnicity_totals = ethnicity_counts.groupby('Study-Title')['Cell count'].sum().reset_index()\n",
    "    ethnicity_totals.rename(columns={'Cell count': 'Total Count'}, inplace=True)\n",
    "    ethnicity_counts = ethnicity_counts.merge(ethnicity_totals, on='Study-Title')\n",
    "    ethnicity_counts['Percentage'] = (ethnicity_counts['Cell count'] / ethnicity_counts['Total Count']) * 100\n",
    "\n",
    "    # Creating a bar plot for percentages by ethnicity\n",
    "    fig_percentage = px.bar(\n",
    "        ethnicity_counts, \n",
    "        x='Study-Title', \n",
    "        y='Percentage', \n",
    "        color='Ethnicity',\n",
    "        labels={'Percentage': 'Percentage (%)', 'Ethnicity': 'Ethnicity'},\n",
    "        title=\"Cell Counts Percentage by Ethnicity\",\n",
    "        text='Percentage'\n",
    "    )\n",
    "    fig_percentage.update_traces(texttemplate='%{text:.2f}%', textposition='outside')\n",
    "    fig_percentage.update_layout(\n",
    "        yaxis_title='Percentage (%)',\n",
    "        title_x=0.5\n",
    "    )\n",
    "\n",
    "    # Aggregating the overall cell counts by ethnicity for the pie chart\n",
    "    total_ethnicity_counts = filtered_df.groupby('Ethnicity')['Cell count'].sum().reset_index()\n",
    "    total_ethnicity_counts.rename(columns={'Cell count': 'Total Count'}, inplace=True)\n",
    "\n",
    "    # Creating a pie chart for the overall distribution of cell counts by ethnicity\n",
    "    pie_fig = px.pie(\n",
    "        total_ethnicity_counts, \n",
    "        names='Ethnicity', \n",
    "        values='Total Count',\n",
    "        labels={'Total Count': 'Number of Cells', 'Ethnicity': 'Ethnicity'},\n",
    "        title=\"Overall Distribution of Cells by Ethnicity\",\n",
    "        hole=0.3\n",
    "    )\n",
    "    pie_fig.update_traces(textinfo='percent+label')\n",
    "    pie_fig.update_layout(title_x=0.5)\n",
    "\n",
    "    return fig_counts, fig_percentage, pie_fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Callback for treemap with ethnicity and total cell counts\n",
    "@app.callback(\n",
    "    [Output('treemap-ethnicity', 'figure'),\n",
    "     Output('treemap-total', 'figure')],\n",
    "    [Input('disease-dropdown', 'value'),\n",
    "     Input('Tissue-Broad-dropdown', 'value'),\n",
    "     Input('modality-dropdown', 'value'),\n",
    "     Input('repository-dropdown', 'value')]\n",
    ")\n",
    "def update_treemaps(selected_diseases, selected_tissue, selected_modalities, selected_repositories):\n",
    "    # Ensure inputs are lists for consistent filtering\n",
    "    selected_diseases = selected_diseases if isinstance(selected_diseases, list) else [selected_diseases]\n",
    "    selected_tissue = selected_tissue if isinstance(selected_tissue, list) else [selected_tissue]\n",
    "    selected_modalities = selected_modalities if isinstance(selected_modalities, list) else [selected_modalities]\n",
    "    selected_repositories = selected_repositories if isinstance(selected_repositories, list) else [selected_repositories]\n",
    "\n",
    "    # Copy the DataFrame and fill NaN values for filtering\n",
    "    df_copy = df.copy()\n",
    "    df_copy['Disease'] = df_copy['Disease'].fillna('')\n",
    "    df_copy['Tissue-Broad'] = df_copy['Tissue-Broad'].fillna('')\n",
    "    df_copy['Data-modality'] = df_copy['Data-modality'].fillna('')\n",
    "    df_copy['Repository'] = df_copy['Repository'].fillna('')\n",
    "    df_copy['Ethnicity'] = df_copy['Ethnicity'].fillna('')\n",
    "\n",
    "    # Apply conditional filtering based on selected dropdown values\n",
    "    filtered_df = df_copy[\n",
    "        (df_copy['Disease'].apply(lambda x: any_in_string(selected_diseases, x)) if selected_diseases else True) &\n",
    "        (df_copy['Tissue-Broad'].apply(lambda x: any_in_string(selected_tissue, x)) if selected_tissue else True) &\n",
    "        (df_copy['Data-modality'].apply(lambda x: any_in_string(selected_modalities, x)) if selected_modalities else True) &\n",
    "        (df_copy['Repository'].apply(lambda x: any_in_string(selected_repositories, x)) if selected_repositories else True)\n",
    "    ]\n",
    "\n",
    "    if filtered_df.empty:\n",
    "        # Create an empty figure with a message when no data is available\n",
    "        fig_ethnicity = px.treemap(title=\"No data available for selected filters.\")\n",
    "        fig_total = px.treemap(title=\"No data available for selected filters.\")\n",
    "    else:\n",
    "        # Treemap based on Ethnicity\n",
    "        fig_ethnicity = px.treemap(\n",
    "            filtered_df,\n",
    "            path=['Repository', 'Ethnicity'],\n",
    "            values='Cell count',\n",
    "            title='Treemap of Heart Data Consortium by Ethnicity'\n",
    "        )\n",
    "\n",
    "        # Treemap for all cell counts (showing repository and study title hierarchy)\n",
    "        fig_total = px.treemap(\n",
    "            filtered_df,\n",
    "            path=['Repository', 'Study-Title'],\n",
    "            values='Cell count',\n",
    "            title='Treemap of Heart Data Consortium by Total Cell Counts'\n",
    "        )\n",
    "\n",
    "    # Update the traces to customize the text and position for ethnicity treemap\n",
    "    fig_ethnicity.update_traces(\n",
    "        texttemplate='%{label}<br>Cell count: %{value}',\n",
    "        textposition='middle center'\n",
    "    )\n",
    "\n",
    "    # Update the traces for the total cell counts treemap\n",
    "    fig_total.update_traces(\n",
    "        texttemplate='%{label}<br>Cell count: %{value}',\n",
    "        textposition='middle center'\n",
    "    )\n",
    "\n",
    "    # Center the titles for both figures\n",
    "    fig_ethnicity.update_layout(title_x=0.5)\n",
    "    fig_total.update_layout(title_x=0.5)\n",
    "\n",
    "    return fig_ethnicity, fig_total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"100%\"\n",
       "            height=\"650\"\n",
       "            src=\"http://127.0.0.1:8051/\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7f9ca3fbb620>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Run the app\n",
    "if __name__ == '__main__':\n",
    "    app.run_server(debug=True, port=8051)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
